In [3]:
import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Part A

#### Build a baseline model without predictor normalization

In [84]:
data = pd.read_csv('https://cocl.us/concrete_data')
data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [85]:
# Splitting the data into predictors and target

predictors = data[data.columns[data.columns != 'Strength']]
target = data['Strength']
print(target.shape)

(1030,)


In [86]:
#Let's save the number of predictors to n_cols since we will need this number when building our network.

n_cols = predictors.shape[1]
n_cols

8

### Use the Keras library to build a neural network with the following:
 
- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

In [87]:
model = tf.keras.Sequential([
                        tf.keras.layers.Dense(10,activation="relu",input_shape=(n_cols,)),
                        tf.keras.layers.Dense(1)
                        ])

model.compile(optimizer='adam', loss='mean_squared_error')

In [88]:
mse_list = []
for i in range(1,51):
    
    # Randomly split the data into a training and test sets by holding 30% of the data for testing.
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(predictors,target,test_size=0.3)
    
    # Fit the model and train for 50 epochs
    model.fit(X_train, y_train, epochs=50, verbose=2)
    
    # Predict the target
    y_pred = model.predict(X_test)
    
    # Calculate mean squared error and store it in a list
    from sklearn.metrics import mean_squared_error
    mse = mean_squared_error(y_test,y_pred)
    mse_list.append(mse)

Train on 721 samples
Epoch 1/50
721/721 - 1s - loss: 216685.5028
Epoch 2/50
721/721 - 0s - loss: 114649.4529
Epoch 3/50
721/721 - 0s - loss: 61996.0313
Epoch 4/50
721/721 - 0s - loss: 30287.1703
Epoch 5/50
721/721 - 0s - loss: 10326.4470
Epoch 6/50
721/721 - 0s - loss: 3792.8639
Epoch 7/50
721/721 - 0s - loss: 3049.7350
Epoch 8/50
721/721 - 0s - loss: 2821.8617
Epoch 9/50
721/721 - 0s - loss: 2595.2441
Epoch 10/50
721/721 - 0s - loss: 2404.1218
Epoch 11/50
721/721 - 0s - loss: 2224.7984
Epoch 12/50
721/721 - 0s - loss: 2059.4909
Epoch 13/50
721/721 - 0s - loss: 1911.1951
Epoch 14/50
721/721 - 0s - loss: 1772.3948
Epoch 15/50
721/721 - 0s - loss: 1634.4435
Epoch 16/50
721/721 - 0s - loss: 1509.8143
Epoch 17/50
721/721 - 0s - loss: 1390.1820
Epoch 18/50
721/721 - 0s - loss: 1281.5344
Epoch 19/50
721/721 - 0s - loss: 1178.9413
Epoch 20/50
721/721 - 0s - loss: 1089.3706
Epoch 21/50
721/721 - 0s - loss: 1001.5922
Epoch 22/50
721/721 - 0s - loss: 924.3927
Epoch 23/50
721/721 - 0s - loss: 849

Epoch 46/50
721/721 - 0s - loss: 90.4265
Epoch 47/50
721/721 - 0s - loss: 93.0547
Epoch 48/50
721/721 - 0s - loss: 91.3509
Epoch 49/50
721/721 - 0s - loss: 90.3838
Epoch 50/50
721/721 - 0s - loss: 92.7431
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 89.7846
Epoch 2/50
721/721 - 0s - loss: 87.3191
Epoch 3/50
721/721 - 0s - loss: 85.7745
Epoch 4/50
721/721 - 0s - loss: 86.6933
Epoch 5/50
721/721 - 0s - loss: 85.1372
Epoch 6/50
721/721 - 0s - loss: 85.6604
Epoch 7/50
721/721 - 0s - loss: 85.7444
Epoch 8/50
721/721 - 0s - loss: 85.1456
Epoch 9/50
721/721 - 0s - loss: 84.2566
Epoch 10/50
721/721 - 0s - loss: 87.7716
Epoch 11/50
721/721 - 0s - loss: 91.2033
Epoch 12/50
721/721 - 0s - loss: 83.7060
Epoch 13/50
721/721 - 0s - loss: 83.3846
Epoch 14/50
721/721 - 0s - loss: 85.3621
Epoch 15/50
721/721 - 0s - loss: 83.0727
Epoch 16/50
721/721 - 0s - loss: 84.1820
Epoch 17/50
721/721 - 0s - loss: 86.5408
Epoch 18/50
721/721 - 0s - loss: 83.8794
Epoch 19/50
721/721 - 0s - loss: 82.5669
Epoc

Epoch 45/50
721/721 - 0s - loss: 59.3918
Epoch 46/50
721/721 - 0s - loss: 60.6592
Epoch 47/50
721/721 - 0s - loss: 58.9377
Epoch 48/50
721/721 - 0s - loss: 57.8557
Epoch 49/50
721/721 - 0s - loss: 55.0478
Epoch 50/50
721/721 - 0s - loss: 54.6585
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 59.7981
Epoch 2/50
721/721 - 0s - loss: 58.2228
Epoch 3/50
721/721 - 0s - loss: 58.5575
Epoch 4/50
721/721 - 0s - loss: 59.0931
Epoch 5/50
721/721 - 0s - loss: 60.6031
Epoch 6/50
721/721 - 0s - loss: 56.4078
Epoch 7/50
721/721 - 0s - loss: 55.9810
Epoch 8/50
721/721 - 0s - loss: 55.4717
Epoch 9/50
721/721 - 0s - loss: 58.7700
Epoch 10/50
721/721 - 0s - loss: 56.0653
Epoch 11/50
721/721 - 0s - loss: 55.5737
Epoch 12/50
721/721 - 0s - loss: 55.4017
Epoch 13/50
721/721 - 0s - loss: 56.2601
Epoch 14/50
721/721 - 0s - loss: 55.9715
Epoch 15/50
721/721 - 0s - loss: 54.1843
Epoch 16/50
721/721 - 0s - loss: 54.4744
Epoch 17/50
721/721 - 0s - loss: 54.3226
Epoch 18/50
721/721 - 0s - loss: 52.9181
Epoc

Epoch 44/50
721/721 - 0s - loss: 48.8495
Epoch 45/50
721/721 - 0s - loss: 47.2670
Epoch 46/50
721/721 - 0s - loss: 50.0903
Epoch 47/50
721/721 - 0s - loss: 48.9497
Epoch 48/50
721/721 - 0s - loss: 48.0895
Epoch 49/50
721/721 - 0s - loss: 48.5163
Epoch 50/50
721/721 - 0s - loss: 48.4230
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 45.6999
Epoch 2/50
721/721 - 0s - loss: 45.4444
Epoch 3/50
721/721 - 0s - loss: 45.4899
Epoch 4/50
721/721 - 0s - loss: 45.8896
Epoch 5/50
721/721 - 0s - loss: 48.0794
Epoch 6/50
721/721 - 0s - loss: 48.5536
Epoch 7/50
721/721 - 0s - loss: 45.8065
Epoch 8/50
721/721 - 0s - loss: 47.6635
Epoch 9/50
721/721 - 0s - loss: 47.2252
Epoch 10/50
721/721 - 0s - loss: 45.1240
Epoch 11/50
721/721 - 0s - loss: 45.1129
Epoch 12/50
721/721 - 0s - loss: 46.0726
Epoch 13/50
721/721 - 0s - loss: 46.2738
Epoch 14/50
721/721 - 0s - loss: 45.1212
Epoch 15/50
721/721 - 0s - loss: 47.5833
Epoch 16/50
721/721 - 0s - loss: 45.9712
Epoch 17/50
721/721 - 0s - loss: 47.0581
Epoc

Epoch 43/50
721/721 - 0s - loss: 47.2196
Epoch 44/50
721/721 - 0s - loss: 46.3403
Epoch 45/50
721/721 - 0s - loss: 47.3591
Epoch 46/50
721/721 - 0s - loss: 48.0989
Epoch 47/50
721/721 - 0s - loss: 47.0639
Epoch 48/50
721/721 - 0s - loss: 47.7703
Epoch 49/50
721/721 - 0s - loss: 50.7348
Epoch 50/50
721/721 - 0s - loss: 50.1369
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 46.2757
Epoch 2/50
721/721 - 0s - loss: 46.2662
Epoch 3/50
721/721 - 0s - loss: 47.2005
Epoch 4/50
721/721 - 0s - loss: 45.8690
Epoch 5/50
721/721 - 0s - loss: 46.9841
Epoch 6/50
721/721 - 0s - loss: 46.2259
Epoch 7/50
721/721 - 0s - loss: 44.7907
Epoch 8/50
721/721 - 0s - loss: 48.6654
Epoch 9/50
721/721 - 0s - loss: 50.1411
Epoch 10/50
721/721 - 0s - loss: 46.4845
Epoch 11/50
721/721 - 0s - loss: 50.7954
Epoch 12/50
721/721 - 0s - loss: 46.7698
Epoch 13/50
721/721 - 0s - loss: 47.3176
Epoch 14/50
721/721 - 0s - loss: 46.6292
Epoch 15/50
721/721 - 0s - loss: 45.1462
Epoch 16/50
721/721 - 0s - loss: 46.4279
Epoc

Epoch 42/50
721/721 - 0s - loss: 45.6393
Epoch 43/50
721/721 - 0s - loss: 45.2615
Epoch 44/50
721/721 - 0s - loss: 45.6909
Epoch 45/50
721/721 - 0s - loss: 46.3853
Epoch 46/50
721/721 - 0s - loss: 46.0853
Epoch 47/50
721/721 - 0s - loss: 44.2050
Epoch 48/50
721/721 - 0s - loss: 47.6330
Epoch 49/50
721/721 - 0s - loss: 44.8165
Epoch 50/50
721/721 - 0s - loss: 44.4758
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 46.3747
Epoch 2/50
721/721 - 0s - loss: 46.5217
Epoch 3/50
721/721 - 0s - loss: 44.3395
Epoch 4/50
721/721 - 0s - loss: 42.8390
Epoch 5/50
721/721 - 0s - loss: 43.6072
Epoch 6/50
721/721 - 0s - loss: 42.3803
Epoch 7/50
721/721 - 0s - loss: 42.4912
Epoch 8/50
721/721 - 0s - loss: 42.8760
Epoch 9/50
721/721 - 0s - loss: 44.8669
Epoch 10/50
721/721 - 0s - loss: 43.7197
Epoch 11/50
721/721 - 0s - loss: 44.0665
Epoch 12/50
721/721 - 0s - loss: 44.8256
Epoch 13/50
721/721 - 0s - loss: 43.5818
Epoch 14/50
721/721 - 0s - loss: 43.8978
Epoch 15/50
721/721 - 0s - loss: 44.2538
Epoc

Epoch 41/50
721/721 - 0s - loss: 44.6328
Epoch 42/50
721/721 - 0s - loss: 48.4903
Epoch 43/50
721/721 - 0s - loss: 50.3543
Epoch 44/50
721/721 - 0s - loss: 46.9304
Epoch 45/50
721/721 - 0s - loss: 45.2137
Epoch 46/50
721/721 - 0s - loss: 44.8793
Epoch 47/50
721/721 - 0s - loss: 44.7509
Epoch 48/50
721/721 - 0s - loss: 45.1716
Epoch 49/50
721/721 - 0s - loss: 45.5058
Epoch 50/50
721/721 - 0s - loss: 45.8558
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 46.4369
Epoch 2/50
721/721 - 0s - loss: 49.4245
Epoch 3/50
721/721 - 0s - loss: 48.1840
Epoch 4/50
721/721 - 0s - loss: 47.5590
Epoch 5/50
721/721 - 0s - loss: 45.8576
Epoch 6/50
721/721 - 0s - loss: 48.1703
Epoch 7/50
721/721 - 0s - loss: 46.8874
Epoch 8/50
721/721 - 0s - loss: 56.0717
Epoch 9/50
721/721 - 0s - loss: 45.9080
Epoch 10/50
721/721 - 0s - loss: 46.6475
Epoch 11/50
721/721 - 0s - loss: 46.4565
Epoch 12/50
721/721 - 0s - loss: 47.8749
Epoch 13/50
721/721 - 0s - loss: 45.7033
Epoch 14/50
721/721 - 0s - loss: 50.5947
Epoc

Epoch 40/50
721/721 - 0s - loss: 48.3807
Epoch 41/50
721/721 - 0s - loss: 47.1244
Epoch 42/50
721/721 - 0s - loss: 45.6738
Epoch 43/50
721/721 - 0s - loss: 45.2045
Epoch 44/50
721/721 - 0s - loss: 44.9052
Epoch 45/50
721/721 - 0s - loss: 44.6025
Epoch 46/50
721/721 - 0s - loss: 46.3419
Epoch 47/50
721/721 - 0s - loss: 45.5160
Epoch 48/50
721/721 - 0s - loss: 44.3152
Epoch 49/50
721/721 - 0s - loss: 44.0564
Epoch 50/50
721/721 - 0s - loss: 45.3624
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 46.1997
Epoch 2/50
721/721 - 0s - loss: 49.5906
Epoch 3/50
721/721 - 0s - loss: 46.5597
Epoch 4/50
721/721 - 0s - loss: 46.3600
Epoch 5/50
721/721 - 0s - loss: 47.2514
Epoch 6/50
721/721 - 0s - loss: 45.8282
Epoch 7/50
721/721 - 0s - loss: 44.9442
Epoch 8/50
721/721 - 0s - loss: 45.1222
Epoch 9/50
721/721 - 0s - loss: 45.3916
Epoch 10/50
721/721 - 0s - loss: 46.5349
Epoch 11/50
721/721 - 0s - loss: 47.4886
Epoch 12/50
721/721 - 0s - loss: 47.9666
Epoch 13/50
721/721 - 0s - loss: 46.1515
Epoc

Epoch 39/50
721/721 - 0s - loss: 43.9961
Epoch 40/50
721/721 - 0s - loss: 45.9013
Epoch 41/50
721/721 - 0s - loss: 43.7586
Epoch 42/50
721/721 - 0s - loss: 44.4278
Epoch 43/50
721/721 - 0s - loss: 43.7504
Epoch 44/50
721/721 - 0s - loss: 43.6794
Epoch 45/50
721/721 - 0s - loss: 43.2459
Epoch 46/50
721/721 - 0s - loss: 44.4178
Epoch 47/50
721/721 - 0s - loss: 43.8588
Epoch 48/50
721/721 - 0s - loss: 43.1354
Epoch 49/50
721/721 - 0s - loss: 44.0120
Epoch 50/50
721/721 - 0s - loss: 43.9880
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 43.6314
Epoch 2/50
721/721 - 0s - loss: 42.5676
Epoch 3/50
721/721 - 0s - loss: 42.0865
Epoch 4/50
721/721 - 0s - loss: 42.1630
Epoch 5/50
721/721 - 0s - loss: 44.8360
Epoch 6/50
721/721 - 0s - loss: 42.0076
Epoch 7/50
721/721 - 0s - loss: 42.1009
Epoch 8/50
721/721 - 0s - loss: 43.3658
Epoch 9/50
721/721 - 0s - loss: 43.9501
Epoch 10/50
721/721 - 0s - loss: 42.9337
Epoch 11/50
721/721 - 0s - loss: 43.7169
Epoch 12/50
721/721 - 0s - loss: 41.6762
Epoc

Epoch 38/50
721/721 - 0s - loss: 44.2297
Epoch 39/50
721/721 - 0s - loss: 43.2173
Epoch 40/50
721/721 - 0s - loss: 43.2647
Epoch 41/50
721/721 - 0s - loss: 51.5396
Epoch 42/50
721/721 - 0s - loss: 42.7322
Epoch 43/50
721/721 - 0s - loss: 43.2277
Epoch 44/50
721/721 - 0s - loss: 44.3773
Epoch 45/50
721/721 - 0s - loss: 45.1911
Epoch 46/50
721/721 - 0s - loss: 42.6626
Epoch 47/50
721/721 - 0s - loss: 43.3847
Epoch 48/50
721/721 - 0s - loss: 42.8192
Epoch 49/50
721/721 - 0s - loss: 42.9893
Epoch 50/50
721/721 - 0s - loss: 42.9055
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 43.7426
Epoch 2/50
721/721 - 0s - loss: 43.2346
Epoch 3/50
721/721 - 0s - loss: 43.2547
Epoch 4/50
721/721 - 0s - loss: 45.1731
Epoch 5/50
721/721 - 0s - loss: 43.4566
Epoch 6/50
721/721 - 0s - loss: 43.6300
Epoch 7/50
721/721 - 0s - loss: 44.3206
Epoch 8/50
721/721 - 0s - loss: 43.9325
Epoch 9/50
721/721 - 0s - loss: 43.7465
Epoch 10/50
721/721 - 0s - loss: 45.6750
Epoch 11/50
721/721 - 0s - loss: 44.3002
Epoc

Epoch 37/50
721/721 - 0s - loss: 39.1770
Epoch 38/50
721/721 - 0s - loss: 40.1589
Epoch 39/50
721/721 - 0s - loss: 40.6758
Epoch 40/50
721/721 - 0s - loss: 41.0137
Epoch 41/50
721/721 - 0s - loss: 41.6491
Epoch 42/50
721/721 - 0s - loss: 40.5399
Epoch 43/50
721/721 - 0s - loss: 40.3407
Epoch 44/50
721/721 - 0s - loss: 40.1525
Epoch 45/50
721/721 - 0s - loss: 39.5800
Epoch 46/50
721/721 - 0s - loss: 40.2948
Epoch 47/50
721/721 - 0s - loss: 41.3470
Epoch 48/50
721/721 - 0s - loss: 40.8532
Epoch 49/50
721/721 - 0s - loss: 40.3414
Epoch 50/50
721/721 - 0s - loss: 39.3773
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 38.9280
Epoch 2/50
721/721 - 0s - loss: 38.9349
Epoch 3/50
721/721 - 0s - loss: 38.7427
Epoch 4/50
721/721 - 0s - loss: 40.2917
Epoch 5/50
721/721 - 0s - loss: 40.4945
Epoch 6/50
721/721 - 0s - loss: 41.6129
Epoch 7/50
721/721 - 0s - loss: 38.5030
Epoch 8/50
721/721 - 0s - loss: 38.9559
Epoch 9/50
721/721 - 0s - loss: 38.5543
Epoch 10/50
721/721 - 0s - loss: 38.5598
Epoc

Epoch 36/50
721/721 - 0s - loss: 39.9187
Epoch 37/50
721/721 - 0s - loss: 39.3432
Epoch 38/50
721/721 - 0s - loss: 39.2342
Epoch 39/50
721/721 - 0s - loss: 39.9842
Epoch 40/50
721/721 - 0s - loss: 41.5027
Epoch 41/50
721/721 - 0s - loss: 41.1604
Epoch 42/50
721/721 - 0s - loss: 40.8615
Epoch 43/50
721/721 - 0s - loss: 40.3839
Epoch 44/50
721/721 - 0s - loss: 40.3157
Epoch 45/50
721/721 - 0s - loss: 41.3133
Epoch 46/50
721/721 - 0s - loss: 50.5238
Epoch 47/50
721/721 - 0s - loss: 42.5592
Epoch 48/50
721/721 - 0s - loss: 41.4754
Epoch 49/50
721/721 - 0s - loss: 44.0210
Epoch 50/50
721/721 - 0s - loss: 41.7856
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 42.7995
Epoch 2/50
721/721 - 0s - loss: 40.4792
Epoch 3/50
721/721 - 0s - loss: 40.3302
Epoch 4/50
721/721 - 0s - loss: 39.4032
Epoch 5/50
721/721 - 0s - loss: 40.9171
Epoch 6/50
721/721 - 0s - loss: 41.2573
Epoch 7/50
721/721 - 0s - loss: 39.4879
Epoch 8/50
721/721 - 0s - loss: 40.2575
Epoch 9/50
721/721 - 0s - loss: 40.2427
Epoc

Epoch 35/50
721/721 - 0s - loss: 39.4112
Epoch 36/50
721/721 - 0s - loss: 40.2745
Epoch 37/50
721/721 - 0s - loss: 41.7880
Epoch 38/50
721/721 - 0s - loss: 41.4843
Epoch 39/50
721/721 - 0s - loss: 39.2625
Epoch 40/50
721/721 - 0s - loss: 39.5327
Epoch 41/50
721/721 - 0s - loss: 39.6049
Epoch 42/50
721/721 - 0s - loss: 42.4558
Epoch 43/50
721/721 - 0s - loss: 39.5507
Epoch 44/50
721/721 - 0s - loss: 39.6397
Epoch 45/50
721/721 - 0s - loss: 39.5105
Epoch 46/50
721/721 - 0s - loss: 39.4722
Epoch 47/50
721/721 - 0s - loss: 38.8994
Epoch 48/50
721/721 - 0s - loss: 40.5643
Epoch 49/50
721/721 - 0s - loss: 41.9936
Epoch 50/50
721/721 - 0s - loss: 42.0674
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 38.4884
Epoch 2/50
721/721 - 0s - loss: 40.3608
Epoch 3/50
721/721 - 0s - loss: 38.5263
Epoch 4/50
721/721 - 0s - loss: 38.0967
Epoch 5/50
721/721 - 0s - loss: 38.9048
Epoch 6/50
721/721 - 0s - loss: 38.4621
Epoch 7/50
721/721 - 0s - loss: 38.8041
Epoch 8/50
721/721 - 0s - loss: 38.6156
Epo

In [89]:
print(mse_list)

[217.45748789506013, 114.82863634215876, 101.40692125791878, 98.60069161704591, 96.3057421984008, 79.27754026086608, 67.64075515491724, 62.187851669953595, 46.71907851244465, 49.27268913670551, 48.146140586105794, 42.27727370850179, 47.699921046318536, 58.87400580333154, 58.9582377583202, 42.35964020518385, 45.02992583802132, 44.40405552754815, 51.68355562914117, 46.135598535183746, 53.37159431505976, 41.66236975021517, 42.72838983370944, 42.73372145349694, 40.26329743398692, 45.91857267827159, 45.34599792445648, 42.30765731818868, 42.237251624145294, 43.385213285710044, 48.19830749896626, 45.66186425042885, 50.75769367221991, 43.707042976429264, 44.783716105082, 40.69888293670249, 44.04421266657311, 48.03503892111003, 39.38222677311855, 42.11860693534304, 45.292247196968184, 39.73566541534696, 37.140345461352624, 43.58938876550807, 43.402884132060436, 46.52029841287069, 48.84716859708978, 39.628759099301654, 43.33784951995237, 40.58648886503471]


In [90]:
# Calculate mean and standard deviation for the stored mse values

mse_mean = np.array(mse_list).mean()
mse_std = np.array(mse_list).std()

print(f'The mean of the mean squared errors is {mse_mean}')
print(f'The standard deviation of the mean squared errors is {mse_std}')

The mean of the mean squared errors is 54.693770050036534
The standard deviation of the mean squared errors is 28.804292495928777


## Part B

### Baseline model with Predictor Normalization

In [91]:
data = pd.read_csv('https://cocl.us/concrete_data')
data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [92]:
# Splitting the data into predictors and target

predictors = data[data.columns[data.columns != 'Strength']]
target = data['Strength']
print(target.shape)

(1030,)


In [93]:
# Normalize the data

predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [94]:
#Let's save the number of predictors to n_cols since we will need this number when building our network.

n_cols = predictors_norm.shape[1]
n_cols

8

In [95]:
model = tf.keras.Sequential([
                        tf.keras.layers.Dense(10,activation="relu",input_shape=(n_cols,)),
                        tf.keras.layers.Dense(1)
                        ])

model.compile(optimizer='adam', loss='mean_squared_error')

In [96]:
mse_list = []
for i in range(1,51):
    
    # Randomly split the data into a training and test sets by holding 30% of the data for testing.
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm,target,test_size=0.3)
    
    # Fit the model and train for 50 epochs
    model.fit(X_train, y_train, epochs=50, verbose=2)
    
    # Predict the target
    y_pred = model.predict(X_test)
    
    # Calculate mean squared error and store it in a list
    from sklearn.metrics import mean_squared_error
    mse = mean_squared_error(y_test,y_pred)
    mse_list.append(mse)

Train on 721 samples
Epoch 1/50
721/721 - 1s - loss: 1542.7854
Epoch 2/50
721/721 - 0s - loss: 1526.9274
Epoch 3/50
721/721 - 0s - loss: 1511.0193
Epoch 4/50
721/721 - 0s - loss: 1495.1978
Epoch 5/50
721/721 - 0s - loss: 1478.2386
Epoch 6/50
721/721 - 0s - loss: 1461.0401
Epoch 7/50
721/721 - 0s - loss: 1443.4131
Epoch 8/50
721/721 - 0s - loss: 1424.5695
Epoch 9/50
721/721 - 0s - loss: 1405.1892
Epoch 10/50
721/721 - 0s - loss: 1384.7503
Epoch 11/50
721/721 - 0s - loss: 1363.4585
Epoch 12/50
721/721 - 0s - loss: 1340.8688
Epoch 13/50
721/721 - 0s - loss: 1317.2367
Epoch 14/50
721/721 - 0s - loss: 1292.5316
Epoch 15/50
721/721 - 0s - loss: 1266.3097
Epoch 16/50
721/721 - 0s - loss: 1238.9362
Epoch 17/50
721/721 - 0s - loss: 1209.8121
Epoch 18/50
721/721 - 0s - loss: 1179.4004
Epoch 19/50
721/721 - 0s - loss: 1147.4712
Epoch 20/50
721/721 - 0s - loss: 1114.6442
Epoch 21/50
721/721 - 0s - loss: 1079.9836
Epoch 22/50
721/721 - 0s - loss: 1045.2739
Epoch 23/50
721/721 - 0s - loss: 1008.8994

Epoch 45/50
721/721 - 0s - loss: 94.8730
Epoch 46/50
721/721 - 0s - loss: 94.5665
Epoch 47/50
721/721 - 0s - loss: 94.2541
Epoch 48/50
721/721 - 0s - loss: 94.0065
Epoch 49/50
721/721 - 0s - loss: 93.6998
Epoch 50/50
721/721 - 0s - loss: 93.3239
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 89.7345
Epoch 2/50
721/721 - 0s - loss: 89.0583
Epoch 3/50
721/721 - 0s - loss: 88.4884
Epoch 4/50
721/721 - 0s - loss: 88.0637
Epoch 5/50
721/721 - 0s - loss: 87.6885
Epoch 6/50
721/721 - 0s - loss: 87.2323
Epoch 7/50
721/721 - 0s - loss: 86.7678
Epoch 8/50
721/721 - 0s - loss: 86.3920
Epoch 9/50
721/721 - 0s - loss: 86.0371
Epoch 10/50
721/721 - 0s - loss: 85.6252
Epoch 11/50
721/721 - 0s - loss: 85.2952
Epoch 12/50
721/721 - 0s - loss: 84.9543
Epoch 13/50
721/721 - 0s - loss: 84.4842
Epoch 14/50
721/721 - 0s - loss: 84.1077
Epoch 15/50
721/721 - 0s - loss: 83.6819
Epoch 16/50
721/721 - 0s - loss: 83.2270
Epoch 17/50
721/721 - 0s - loss: 82.8743
Epoch 18/50
721/721 - 0s - loss: 82.4754
Epoc

Epoch 44/50
721/721 - 0s - loss: 44.8674
Epoch 45/50
721/721 - 0s - loss: 44.6955
Epoch 46/50
721/721 - 0s - loss: 44.7096
Epoch 47/50
721/721 - 0s - loss: 44.6319
Epoch 48/50
721/721 - 0s - loss: 44.5035
Epoch 49/50
721/721 - 0s - loss: 44.4851
Epoch 50/50
721/721 - 0s - loss: 44.2847
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 44.6011
Epoch 2/50
721/721 - 0s - loss: 44.3574
Epoch 3/50
721/721 - 0s - loss: 44.2197
Epoch 4/50
721/721 - 0s - loss: 44.0902
Epoch 5/50
721/721 - 0s - loss: 44.0442
Epoch 6/50
721/721 - 0s - loss: 43.8991
Epoch 7/50
721/721 - 0s - loss: 43.8073
Epoch 8/50
721/721 - 0s - loss: 43.7490
Epoch 9/50
721/721 - 0s - loss: 43.6227
Epoch 10/50
721/721 - 0s - loss: 43.5300
Epoch 11/50
721/721 - 0s - loss: 43.4468
Epoch 12/50
721/721 - 0s - loss: 43.4120
Epoch 13/50
721/721 - 0s - loss: 43.2653
Epoch 14/50
721/721 - 0s - loss: 43.1783
Epoch 15/50
721/721 - 0s - loss: 43.1125
Epoch 16/50
721/721 - 0s - loss: 43.0863
Epoch 17/50
721/721 - 0s - loss: 42.9312
Epoc

Epoch 43/50
721/721 - 0s - loss: 37.7251
Epoch 44/50
721/721 - 0s - loss: 37.7827
Epoch 45/50
721/721 - 0s - loss: 37.6956
Epoch 46/50
721/721 - 0s - loss: 37.7691
Epoch 47/50
721/721 - 0s - loss: 37.6606
Epoch 48/50
721/721 - 0s - loss: 37.8437
Epoch 49/50
721/721 - 0s - loss: 37.6889
Epoch 50/50
721/721 - 0s - loss: 37.7205
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 38.9370
Epoch 2/50
721/721 - 0s - loss: 38.7642
Epoch 3/50
721/721 - 0s - loss: 38.6116
Epoch 4/50
721/721 - 0s - loss: 38.6315
Epoch 5/50
721/721 - 0s - loss: 38.6253
Epoch 6/50
721/721 - 0s - loss: 38.4818
Epoch 7/50
721/721 - 0s - loss: 38.4079
Epoch 8/50
721/721 - 0s - loss: 38.3435
Epoch 9/50
721/721 - 0s - loss: 38.3298
Epoch 10/50
721/721 - 0s - loss: 38.3329
Epoch 11/50
721/721 - 0s - loss: 38.2167
Epoch 12/50
721/721 - 0s - loss: 38.2149
Epoch 13/50
721/721 - 0s - loss: 38.1330
Epoch 14/50
721/721 - 0s - loss: 38.1976
Epoch 15/50
721/721 - 0s - loss: 38.0398
Epoch 16/50
721/721 - 0s - loss: 38.0658
Epoc

Epoch 42/50
721/721 - 0s - loss: 35.1569
Epoch 43/50
721/721 - 0s - loss: 35.1575
Epoch 44/50
721/721 - 0s - loss: 35.1645
Epoch 45/50
721/721 - 0s - loss: 35.1614
Epoch 46/50
721/721 - 0s - loss: 35.1374
Epoch 47/50
721/721 - 0s - loss: 35.1478
Epoch 48/50
721/721 - 0s - loss: 35.1800
Epoch 49/50
721/721 - 0s - loss: 35.1106
Epoch 50/50
721/721 - 0s - loss: 35.1407
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 35.9626
Epoch 2/50
721/721 - 0s - loss: 35.7628
Epoch 3/50
721/721 - 0s - loss: 35.6696
Epoch 4/50
721/721 - 0s - loss: 35.6456
Epoch 5/50
721/721 - 0s - loss: 35.6304
Epoch 6/50
721/721 - 0s - loss: 35.5948
Epoch 7/50
721/721 - 0s - loss: 35.5321
Epoch 8/50
721/721 - 0s - loss: 35.5374
Epoch 9/50
721/721 - 0s - loss: 35.4684
Epoch 10/50
721/721 - 0s - loss: 35.5143
Epoch 11/50
721/721 - 0s - loss: 35.4491
Epoch 12/50
721/721 - 0s - loss: 35.4195
Epoch 13/50
721/721 - 0s - loss: 35.4270
Epoch 14/50
721/721 - 0s - loss: 35.4368
Epoch 15/50
721/721 - 0s - loss: 35.3809
Epoc

Epoch 41/50
721/721 - 0s - loss: 33.7555
Epoch 42/50
721/721 - 0s - loss: 33.7153
Epoch 43/50
721/721 - 0s - loss: 33.7178
Epoch 44/50
721/721 - 0s - loss: 33.7221
Epoch 45/50
721/721 - 0s - loss: 33.6797
Epoch 46/50
721/721 - 0s - loss: 33.6999
Epoch 47/50
721/721 - 0s - loss: 33.6558
Epoch 48/50
721/721 - 0s - loss: 33.6978
Epoch 49/50
721/721 - 0s - loss: 33.6250
Epoch 50/50
721/721 - 0s - loss: 33.6425
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 38.3411
Epoch 2/50
721/721 - 0s - loss: 37.9651
Epoch 3/50
721/721 - 0s - loss: 37.8204
Epoch 4/50
721/721 - 0s - loss: 37.6731
Epoch 5/50
721/721 - 0s - loss: 37.5226
Epoch 6/50
721/721 - 0s - loss: 37.4525
Epoch 7/50
721/721 - 0s - loss: 37.3732
Epoch 8/50
721/721 - 0s - loss: 37.3580
Epoch 9/50
721/721 - 0s - loss: 37.4835
Epoch 10/50
721/721 - 0s - loss: 37.3121
Epoch 11/50
721/721 - 0s - loss: 37.2589
Epoch 12/50
721/721 - 0s - loss: 37.2146
Epoch 13/50
721/721 - 0s - loss: 37.1780
Epoch 14/50
721/721 - 0s - loss: 37.1505
Epoc

Epoch 40/50
721/721 - 0s - loss: 34.4669
Epoch 41/50
721/721 - 0s - loss: 34.4591
Epoch 42/50
721/721 - 0s - loss: 34.3793
Epoch 43/50
721/721 - 0s - loss: 34.3873
Epoch 44/50
721/721 - 0s - loss: 34.3568
Epoch 45/50
721/721 - 0s - loss: 34.3631
Epoch 46/50
721/721 - 0s - loss: 34.3943
Epoch 47/50
721/721 - 0s - loss: 34.3124
Epoch 48/50
721/721 - 0s - loss: 34.3744
Epoch 49/50
721/721 - 0s - loss: 34.3161
Epoch 50/50
721/721 - 0s - loss: 34.3198
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 38.0593
Epoch 2/50
721/721 - 0s - loss: 37.8553
Epoch 3/50
721/721 - 0s - loss: 37.8488
Epoch 4/50
721/721 - 0s - loss: 37.7711
Epoch 5/50
721/721 - 0s - loss: 37.7638
Epoch 6/50
721/721 - 0s - loss: 37.6837
Epoch 7/50
721/721 - 0s - loss: 37.6775
Epoch 8/50
721/721 - 0s - loss: 37.7690
Epoch 9/50
721/721 - 0s - loss: 37.6601
Epoch 10/50
721/721 - 0s - loss: 37.6472
Epoch 11/50
721/721 - 0s - loss: 37.5941
Epoch 12/50
721/721 - 0s - loss: 37.6642
Epoch 13/50
721/721 - 0s - loss: 37.5841
Epoc

Epoch 39/50
721/721 - 0s - loss: 36.5676
Epoch 40/50
721/721 - 0s - loss: 36.5285
Epoch 41/50
721/721 - 0s - loss: 36.6214
Epoch 42/50
721/721 - 0s - loss: 36.5918
Epoch 43/50
721/721 - 0s - loss: 36.5146
Epoch 44/50
721/721 - 0s - loss: 36.5417
Epoch 45/50
721/721 - 0s - loss: 36.5341
Epoch 46/50
721/721 - 0s - loss: 36.5464
Epoch 47/50
721/721 - 0s - loss: 36.5393
Epoch 48/50
721/721 - 0s - loss: 36.5743
Epoch 49/50
721/721 - 0s - loss: 36.5572
Epoch 50/50
721/721 - 0s - loss: 36.5018
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 37.0091
Epoch 2/50
721/721 - 0s - loss: 36.8906
Epoch 3/50
721/721 - 0s - loss: 36.8517
Epoch 4/50
721/721 - 0s - loss: 36.8146
Epoch 5/50
721/721 - 0s - loss: 36.8638
Epoch 6/50
721/721 - 0s - loss: 36.7142
Epoch 7/50
721/721 - 0s - loss: 36.8285
Epoch 8/50
721/721 - 0s - loss: 36.7312
Epoch 9/50
721/721 - 0s - loss: 36.7419
Epoch 10/50
721/721 - 0s - loss: 36.6827
Epoch 11/50
721/721 - 0s - loss: 36.6237
Epoch 12/50
721/721 - 0s - loss: 36.6215
Epoc

Epoch 38/50
721/721 - 0s - loss: 35.4696
Epoch 39/50
721/721 - 0s - loss: 35.4258
Epoch 40/50
721/721 - 0s - loss: 35.4275
Epoch 41/50
721/721 - 0s - loss: 35.4422
Epoch 42/50
721/721 - 0s - loss: 35.5265
Epoch 43/50
721/721 - 0s - loss: 35.4191
Epoch 44/50
721/721 - 0s - loss: 35.3687
Epoch 45/50
721/721 - 0s - loss: 35.4461
Epoch 46/50
721/721 - 0s - loss: 35.4281
Epoch 47/50
721/721 - 0s - loss: 35.3582
Epoch 48/50
721/721 - 0s - loss: 35.4271
Epoch 49/50
721/721 - 0s - loss: 35.3532
Epoch 50/50
721/721 - 0s - loss: 35.3590
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 37.8390
Epoch 2/50
721/721 - 0s - loss: 37.6711
Epoch 3/50
721/721 - 0s - loss: 37.6897
Epoch 4/50
721/721 - 0s - loss: 37.5937
Epoch 5/50
721/721 - 0s - loss: 37.5978
Epoch 6/50
721/721 - 0s - loss: 37.6469
Epoch 7/50
721/721 - 0s - loss: 37.4933
Epoch 8/50
721/721 - 0s - loss: 37.5308
Epoch 9/50
721/721 - 0s - loss: 37.5144
Epoch 10/50
721/721 - 0s - loss: 37.4731
Epoch 11/50
721/721 - 0s - loss: 37.4889
Epoc

Epoch 37/50
721/721 - 0s - loss: 33.6582
Epoch 38/50
721/721 - 0s - loss: 33.7067
Epoch 39/50
721/721 - 0s - loss: 33.6222
Epoch 40/50
721/721 - 0s - loss: 33.6573
Epoch 41/50
721/721 - 0s - loss: 33.6745
Epoch 42/50
721/721 - 0s - loss: 33.6358
Epoch 43/50
721/721 - 0s - loss: 33.6388
Epoch 44/50
721/721 - 0s - loss: 33.6144
Epoch 45/50
721/721 - 0s - loss: 33.5858
Epoch 46/50
721/721 - 0s - loss: 33.6085
Epoch 47/50
721/721 - 0s - loss: 33.5753
Epoch 48/50
721/721 - 0s - loss: 33.6490
Epoch 49/50
721/721 - 0s - loss: 33.6004
Epoch 50/50
721/721 - 0s - loss: 33.5617
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 34.2053
Epoch 2/50
721/721 - 0s - loss: 34.1493
Epoch 3/50
721/721 - 0s - loss: 34.0590
Epoch 4/50
721/721 - 0s - loss: 34.0299
Epoch 5/50
721/721 - 0s - loss: 33.9775
Epoch 6/50
721/721 - 0s - loss: 33.9906
Epoch 7/50
721/721 - 0s - loss: 33.9181
Epoch 8/50
721/721 - 0s - loss: 33.9335
Epoch 9/50
721/721 - 0s - loss: 33.9413
Epoch 10/50
721/721 - 0s - loss: 33.8630
Epoc

Epoch 36/50
721/721 - 0s - loss: 34.0565
Epoch 37/50
721/721 - 0s - loss: 34.0583
Epoch 38/50
721/721 - 0s - loss: 34.0703
Epoch 39/50
721/721 - 0s - loss: 34.0863
Epoch 40/50
721/721 - 0s - loss: 34.0168
Epoch 41/50
721/721 - 0s - loss: 34.1093
Epoch 42/50
721/721 - 0s - loss: 34.0304
Epoch 43/50
721/721 - 0s - loss: 34.0562
Epoch 44/50
721/721 - 0s - loss: 34.0122
Epoch 45/50
721/721 - 0s - loss: 34.0714
Epoch 46/50
721/721 - 0s - loss: 34.0566
Epoch 47/50
721/721 - 0s - loss: 34.0850
Epoch 48/50
721/721 - 0s - loss: 34.0417
Epoch 49/50
721/721 - 0s - loss: 34.0660
Epoch 50/50
721/721 - 0s - loss: 34.0856
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 36.8990
Epoch 2/50
721/721 - 0s - loss: 36.8390
Epoch 3/50
721/721 - 0s - loss: 36.6325
Epoch 4/50
721/721 - 0s - loss: 36.6215
Epoch 5/50
721/721 - 0s - loss: 36.6414
Epoch 6/50
721/721 - 0s - loss: 36.5734
Epoch 7/50
721/721 - 0s - loss: 36.5808
Epoch 8/50
721/721 - 0s - loss: 36.5502
Epoch 9/50
721/721 - 0s - loss: 36.5075
Epoc

Epoch 35/50
721/721 - 0s - loss: 32.4814
Epoch 36/50
721/721 - 0s - loss: 32.5470
Epoch 37/50
721/721 - 0s - loss: 32.4749
Epoch 38/50
721/721 - 0s - loss: 32.4467
Epoch 39/50
721/721 - 0s - loss: 32.5395
Epoch 40/50
721/721 - 0s - loss: 32.4514
Epoch 41/50
721/721 - 0s - loss: 32.4819
Epoch 42/50
721/721 - 0s - loss: 32.4435
Epoch 43/50
721/721 - 0s - loss: 32.4381
Epoch 44/50
721/721 - 0s - loss: 32.4211
Epoch 45/50
721/721 - 0s - loss: 32.4402
Epoch 46/50
721/721 - 0s - loss: 32.4893
Epoch 47/50
721/721 - 0s - loss: 32.3976
Epoch 48/50
721/721 - 0s - loss: 32.4142
Epoch 49/50
721/721 - 0s - loss: 32.3895
Epoch 50/50
721/721 - 0s - loss: 32.3633
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 35.5909
Epoch 2/50
721/721 - 0s - loss: 35.3936
Epoch 3/50
721/721 - 0s - loss: 35.2392
Epoch 4/50
721/721 - 0s - loss: 35.1757
Epoch 5/50
721/721 - 0s - loss: 35.1274
Epoch 6/50
721/721 - 0s - loss: 35.0470
Epoch 7/50
721/721 - 0s - loss: 35.0515
Epoch 8/50
721/721 - 0s - loss: 34.9568
Epo

Epoch 34/50
721/721 - 0s - loss: 32.8303
Epoch 35/50
721/721 - 0s - loss: 32.8172
Epoch 36/50
721/721 - 0s - loss: 32.8508
Epoch 37/50
721/721 - 0s - loss: 32.7834
Epoch 38/50
721/721 - 0s - loss: 32.7857
Epoch 39/50
721/721 - 0s - loss: 32.7959
Epoch 40/50
721/721 - 0s - loss: 32.8124
Epoch 41/50
721/721 - 0s - loss: 32.7865
Epoch 42/50
721/721 - 0s - loss: 32.7471
Epoch 43/50
721/721 - 0s - loss: 32.7721
Epoch 44/50
721/721 - 0s - loss: 32.7799
Epoch 45/50
721/721 - 0s - loss: 32.7482
Epoch 46/50
721/721 - 0s - loss: 32.7931
Epoch 47/50
721/721 - 0s - loss: 32.7357
Epoch 48/50
721/721 - 0s - loss: 32.7263
Epoch 49/50
721/721 - 0s - loss: 32.7574
Epoch 50/50
721/721 - 0s - loss: 32.6831
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 34.0864
Epoch 2/50
721/721 - 0s - loss: 33.9990
Epoch 3/50
721/721 - 0s - loss: 33.8595
Epoch 4/50
721/721 - 0s - loss: 33.8030
Epoch 5/50
721/721 - 0s - loss: 33.8027
Epoch 6/50
721/721 - 0s - loss: 33.7481
Epoch 7/50
721/721 - 0s - loss: 33.7478
Ep

In [97]:
print(mse_list)

[300.2801577575382, 157.57988354523332, 105.07330732197893, 93.2943394752134, 82.91336064594952, 69.98532855801206, 45.942762578833424, 49.3769681624087, 48.64534042052291, 42.35486812574144, 39.61691706414029, 43.491947596245545, 42.68446051987104, 37.43256262540845, 37.404398598741366, 42.09102485170716, 42.241676002496774, 39.09782830672964, 41.3317744506081, 45.52399321139057, 35.4753695510808, 36.64138971072175, 36.466243587672, 39.83569817250045, 31.892050913313543, 39.13274602863421, 36.663161654698, 33.380771262862645, 33.32266807425361, 39.39229622912552, 36.058867736321844, 35.39965965730529, 30.617698296609888, 39.928933072759996, 45.717860354737724, 38.80484463266954, 39.38562427487493, 37.95601278251775, 33.072391657361, 37.238730446497165, 32.352337168543855, 36.15019218813994, 38.15195327984751, 41.423448306334755, 35.27075330811555, 36.01186556782207, 42.57735359715568, 39.40367677606954, 37.67857220728793, 33.85924625921677]


In [98]:
# Calculate mean and standard deviation for the stored mse values

mse_mean = np.array(mse_list).mean()
mse_std = np.array(mse_list).std()

print(f'The mean of the mean squared errors is {mse_mean}')
print(f'The standard deviation of the mean squared errors is {mse_std}')

The mean of the mean squared errors is 50.31250633147644
The standard deviation of the mean squared errors is 41.70002554996229


#### The mean of the mean squared errors compared to Part A has decreased from 54.69 to 50.31

## Part C

### Increase the number of epochs to 100

In [99]:
data = pd.read_csv('https://cocl.us/concrete_data')
data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [100]:
# Splitting the data into predictors and target

predictors = data[data.columns[data.columns != 'Strength']]
target = data['Strength']
print(target.shape)

(1030,)


In [101]:
# Normalize the data

predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [102]:
#Let's save the number of predictors to n_cols since we will need this number when building our network.

n_cols = predictors_norm.shape[1]
n_cols

8

In [103]:
model = tf.keras.Sequential([
                        tf.keras.layers.Dense(10,activation="relu",input_shape=(n_cols,)),
                        tf.keras.layers.Dense(1)
                        ])

model.compile(optimizer='adam', loss='mean_squared_error')

In [104]:
mse_list = []
for i in range(1,51):
    
    # Randomly split the data into a training and test sets by holding 30% of the data for testing.
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm,target,test_size=0.3)
    
    # Fit the model and train for 100 epochs
    model.fit(X_train, y_train, epochs=100, verbose=2)
    
    # Predict the target
    y_pred = model.predict(X_test)
    
    # Calculate mean squared error and store it in a list
    from sklearn.metrics import mean_squared_error
    mse = mean_squared_error(y_test,y_pred)
    mse_list.append(mse)

Train on 721 samples
Epoch 1/100
721/721 - 1s - loss: 1571.1305
Epoch 2/100
721/721 - 0s - loss: 1551.7146
Epoch 3/100
721/721 - 0s - loss: 1532.4006
Epoch 4/100
721/721 - 0s - loss: 1513.1183
Epoch 5/100
721/721 - 0s - loss: 1493.7654
Epoch 6/100
721/721 - 0s - loss: 1474.1667
Epoch 7/100
721/721 - 0s - loss: 1454.2882
Epoch 8/100
721/721 - 0s - loss: 1433.7526
Epoch 9/100
721/721 - 0s - loss: 1412.6009
Epoch 10/100
721/721 - 0s - loss: 1391.0963
Epoch 11/100
721/721 - 0s - loss: 1368.7684
Epoch 12/100
721/721 - 0s - loss: 1346.1215
Epoch 13/100
721/721 - 0s - loss: 1323.1098
Epoch 14/100
721/721 - 0s - loss: 1298.9529
Epoch 15/100
721/721 - 0s - loss: 1274.7627
Epoch 16/100
721/721 - 0s - loss: 1249.7598
Epoch 17/100
721/721 - 0s - loss: 1224.4868
Epoch 18/100
721/721 - 0s - loss: 1198.3376
Epoch 19/100
721/721 - 0s - loss: 1171.9956
Epoch 20/100
721/721 - 0s - loss: 1145.1856
Epoch 21/100
721/721 - 0s - loss: 1117.7753
Epoch 22/100
721/721 - 0s - loss: 1089.8537
Epoch 23/100
721/721

Epoch 91/100
721/721 - 0s - loss: 90.5078
Epoch 92/100
721/721 - 0s - loss: 90.1341
Epoch 93/100
721/721 - 0s - loss: 89.7210
Epoch 94/100
721/721 - 0s - loss: 89.3167
Epoch 95/100
721/721 - 0s - loss: 88.9707
Epoch 96/100
721/721 - 0s - loss: 88.5483
Epoch 97/100
721/721 - 0s - loss: 88.2554
Epoch 98/100
721/721 - 0s - loss: 87.7974
Epoch 99/100
721/721 - 0s - loss: 87.5184
Epoch 100/100
721/721 - 0s - loss: 87.1456
Train on 721 samples
Epoch 1/100
721/721 - 0s - loss: 96.4159
Epoch 2/100
721/721 - 0s - loss: 96.0611
Epoch 3/100
721/721 - 0s - loss: 95.4764
Epoch 4/100
721/721 - 0s - loss: 95.1187
Epoch 5/100
721/721 - 0s - loss: 94.5946
Epoch 6/100
721/721 - 0s - loss: 94.2333
Epoch 7/100
721/721 - 0s - loss: 93.8440
Epoch 8/100
721/721 - 0s - loss: 93.4859
Epoch 9/100
721/721 - 0s - loss: 93.0818
Epoch 10/100
721/721 - 0s - loss: 92.7623
Epoch 11/100
721/721 - 0s - loss: 92.4633
Epoch 12/100
721/721 - 0s - loss: 92.0785
Epoch 13/100
721/721 - 0s - loss: 91.7943
Epoch 14/100
721/721 

Epoch 86/100
721/721 - 0s - loss: 44.8744
Epoch 87/100
721/721 - 0s - loss: 44.6838
Epoch 88/100
721/721 - 0s - loss: 44.6193
Epoch 89/100
721/721 - 0s - loss: 44.4776
Epoch 90/100
721/721 - 0s - loss: 44.3083
Epoch 91/100
721/721 - 0s - loss: 44.1448
Epoch 92/100
721/721 - 0s - loss: 44.0559
Epoch 93/100
721/721 - 0s - loss: 43.9263
Epoch 94/100
721/721 - 0s - loss: 43.7953
Epoch 95/100
721/721 - 0s - loss: 43.6300
Epoch 96/100
721/721 - 0s - loss: 43.6266
Epoch 97/100
721/721 - 0s - loss: 43.3389
Epoch 98/100
721/721 - 0s - loss: 43.3331
Epoch 99/100
721/721 - 0s - loss: 43.1932
Epoch 100/100
721/721 - 0s - loss: 43.0914
Train on 721 samples
Epoch 1/100
721/721 - 0s - loss: 46.0595
Epoch 2/100
721/721 - 0s - loss: 45.7748
Epoch 3/100
721/721 - 0s - loss: 45.6637
Epoch 4/100
721/721 - 0s - loss: 45.4640
Epoch 5/100
721/721 - 0s - loss: 45.3529
Epoch 6/100
721/721 - 0s - loss: 45.2293
Epoch 7/100
721/721 - 0s - loss: 45.1369
Epoch 8/100
721/721 - 0s - loss: 45.0182
Epoch 9/100
721/721 

Epoch 81/100
721/721 - 0s - loss: 36.9831
Epoch 82/100
721/721 - 0s - loss: 37.0171
Epoch 83/100
721/721 - 0s - loss: 36.9551
Epoch 84/100
721/721 - 0s - loss: 37.0139
Epoch 85/100
721/721 - 0s - loss: 36.9072
Epoch 86/100
721/721 - 0s - loss: 36.9012
Epoch 87/100
721/721 - 0s - loss: 36.8786
Epoch 88/100
721/721 - 0s - loss: 36.9009
Epoch 89/100
721/721 - 0s - loss: 36.8884
Epoch 90/100
721/721 - 0s - loss: 36.8718
Epoch 91/100
721/721 - 0s - loss: 36.8280
Epoch 92/100
721/721 - 0s - loss: 36.7851
Epoch 93/100
721/721 - 0s - loss: 36.8324
Epoch 94/100
721/721 - 0s - loss: 36.7347
Epoch 95/100
721/721 - 0s - loss: 36.7895
Epoch 96/100
721/721 - 0s - loss: 36.6885
Epoch 97/100
721/721 - 0s - loss: 36.7646
Epoch 98/100
721/721 - 0s - loss: 36.6407
Epoch 99/100
721/721 - 0s - loss: 36.6476
Epoch 100/100
721/721 - 0s - loss: 36.7587
Train on 721 samples
Epoch 1/100
721/721 - 0s - loss: 37.8944
Epoch 2/100
721/721 - 0s - loss: 37.6898
Epoch 3/100
721/721 - 0s - loss: 37.7190
Epoch 4/100
721

Epoch 76/100
721/721 - 0s - loss: 35.3459
Epoch 77/100
721/721 - 0s - loss: 35.3520
Epoch 78/100
721/721 - 0s - loss: 35.2837
Epoch 79/100
721/721 - 0s - loss: 35.3067
Epoch 80/100
721/721 - 0s - loss: 35.3606
Epoch 81/100
721/721 - 0s - loss: 35.3999
Epoch 82/100
721/721 - 0s - loss: 35.3239
Epoch 83/100
721/721 - 0s - loss: 35.2423
Epoch 84/100
721/721 - 0s - loss: 35.2587
Epoch 85/100
721/721 - 0s - loss: 35.2013
Epoch 86/100
721/721 - 0s - loss: 35.2069
Epoch 87/100
721/721 - 0s - loss: 35.2135
Epoch 88/100
721/721 - 0s - loss: 35.1506
Epoch 89/100
721/721 - 0s - loss: 35.1682
Epoch 90/100
721/721 - 0s - loss: 35.2518
Epoch 91/100
721/721 - 0s - loss: 35.1820
Epoch 92/100
721/721 - 0s - loss: 35.1429
Epoch 93/100
721/721 - 0s - loss: 35.1820
Epoch 94/100
721/721 - 0s - loss: 35.1587
Epoch 95/100
721/721 - 0s - loss: 35.2628
Epoch 96/100
721/721 - 0s - loss: 35.1183
Epoch 97/100
721/721 - 0s - loss: 35.1039
Epoch 98/100
721/721 - 0s - loss: 35.0646
Epoch 99/100
721/721 - 0s - loss: 

Epoch 71/100
721/721 - 0s - loss: 31.0876
Epoch 72/100
721/721 - 0s - loss: 31.0877
Epoch 73/100
721/721 - 0s - loss: 31.1704
Epoch 74/100
721/721 - 0s - loss: 31.0808
Epoch 75/100
721/721 - 0s - loss: 31.0930
Epoch 76/100
721/721 - 0s - loss: 31.0392
Epoch 77/100
721/721 - 0s - loss: 31.0680
Epoch 78/100
721/721 - 0s - loss: 31.0800
Epoch 79/100
721/721 - 0s - loss: 31.0480
Epoch 80/100
721/721 - 0s - loss: 31.0359
Epoch 81/100
721/721 - 0s - loss: 31.0031
Epoch 82/100
721/721 - 0s - loss: 31.0681
Epoch 83/100
721/721 - 0s - loss: 30.9742
Epoch 84/100
721/721 - 0s - loss: 30.9822
Epoch 85/100
721/721 - 0s - loss: 30.9423
Epoch 86/100
721/721 - 0s - loss: 30.9442
Epoch 87/100
721/721 - 0s - loss: 31.0728
Epoch 88/100
721/721 - 0s - loss: 31.0606
Epoch 89/100
721/721 - 0s - loss: 31.0891
Epoch 90/100
721/721 - 0s - loss: 30.9203
Epoch 91/100
721/721 - 0s - loss: 30.9815
Epoch 92/100
721/721 - 0s - loss: 30.9087
Epoch 93/100
721/721 - 0s - loss: 30.9273
Epoch 94/100
721/721 - 0s - loss: 

Epoch 66/100
721/721 - 0s - loss: 32.8055
Epoch 67/100
721/721 - 0s - loss: 32.7374
Epoch 68/100
721/721 - 0s - loss: 32.7073
Epoch 69/100
721/721 - 0s - loss: 32.7321
Epoch 70/100
721/721 - 0s - loss: 32.6614
Epoch 71/100
721/721 - 0s - loss: 32.6723
Epoch 72/100
721/721 - 0s - loss: 32.6226
Epoch 73/100
721/721 - 0s - loss: 32.7399
Epoch 74/100
721/721 - 0s - loss: 32.6155
Epoch 75/100
721/721 - 0s - loss: 32.6584
Epoch 76/100
721/721 - 0s - loss: 32.6516
Epoch 77/100
721/721 - 0s - loss: 32.6527
Epoch 78/100
721/721 - 0s - loss: 32.6348
Epoch 79/100
721/721 - 0s - loss: 32.6596
Epoch 80/100
721/721 - 0s - loss: 32.6053
Epoch 81/100
721/721 - 0s - loss: 32.5805
Epoch 82/100
721/721 - 0s - loss: 32.6151
Epoch 83/100
721/721 - 0s - loss: 32.6361
Epoch 84/100
721/721 - 0s - loss: 32.5958
Epoch 85/100
721/721 - 0s - loss: 32.6567
Epoch 86/100
721/721 - 0s - loss: 32.5151
Epoch 87/100
721/721 - 0s - loss: 32.5401
Epoch 88/100
721/721 - 0s - loss: 32.5028
Epoch 89/100
721/721 - 0s - loss: 

Epoch 61/100
721/721 - 0s - loss: 31.5472
Epoch 62/100
721/721 - 0s - loss: 31.5266
Epoch 63/100
721/721 - 0s - loss: 31.6057
Epoch 64/100
721/721 - 0s - loss: 31.6223
Epoch 65/100
721/721 - 0s - loss: 31.5146
Epoch 66/100
721/721 - 0s - loss: 31.5320
Epoch 67/100
721/721 - 0s - loss: 31.5268
Epoch 68/100
721/721 - 0s - loss: 31.4773
Epoch 69/100
721/721 - 0s - loss: 31.5316
Epoch 70/100
721/721 - 0s - loss: 31.4911
Epoch 71/100
721/721 - 0s - loss: 31.4980
Epoch 72/100
721/721 - 0s - loss: 31.5011
Epoch 73/100
721/721 - 0s - loss: 31.5463
Epoch 74/100
721/721 - 0s - loss: 31.4956
Epoch 75/100
721/721 - 0s - loss: 31.4918
Epoch 76/100
721/721 - 0s - loss: 31.4415
Epoch 77/100
721/721 - 0s - loss: 31.4988
Epoch 78/100
721/721 - 0s - loss: 31.4594
Epoch 79/100
721/721 - 0s - loss: 31.4948
Epoch 80/100
721/721 - 0s - loss: 31.4280
Epoch 81/100
721/721 - 0s - loss: 31.4003
Epoch 82/100
721/721 - 0s - loss: 31.4659
Epoch 83/100
721/721 - 0s - loss: 31.4879
Epoch 84/100
721/721 - 0s - loss: 

Epoch 56/100
721/721 - 0s - loss: 27.7892
Epoch 57/100
721/721 - 0s - loss: 28.0329
Epoch 58/100
721/721 - 0s - loss: 27.7779
Epoch 59/100
721/721 - 0s - loss: 27.8153
Epoch 60/100
721/721 - 0s - loss: 27.7954
Epoch 61/100
721/721 - 0s - loss: 27.7745
Epoch 62/100
721/721 - 0s - loss: 27.7602
Epoch 63/100
721/721 - 0s - loss: 27.7347
Epoch 64/100
721/721 - 0s - loss: 27.7635
Epoch 65/100
721/721 - 0s - loss: 27.7958
Epoch 66/100
721/721 - 0s - loss: 27.7818
Epoch 67/100
721/721 - 0s - loss: 27.7027
Epoch 68/100
721/721 - 0s - loss: 27.7193
Epoch 69/100
721/721 - 0s - loss: 27.7426
Epoch 70/100
721/721 - 0s - loss: 27.7016
Epoch 71/100
721/721 - 0s - loss: 27.6929
Epoch 72/100
721/721 - 0s - loss: 27.7404
Epoch 73/100
721/721 - 0s - loss: 27.7301
Epoch 74/100
721/721 - 0s - loss: 27.7187
Epoch 75/100
721/721 - 0s - loss: 27.6736
Epoch 76/100
721/721 - 0s - loss: 27.6666
Epoch 77/100
721/721 - 0s - loss: 27.6613
Epoch 78/100
721/721 - 0s - loss: 27.6327
Epoch 79/100
721/721 - 0s - loss: 

Epoch 51/100
721/721 - 0s - loss: 30.4704
Epoch 52/100
721/721 - 0s - loss: 30.4895
Epoch 53/100
721/721 - 0s - loss: 30.4752
Epoch 54/100
721/721 - 0s - loss: 30.4214
Epoch 55/100
721/721 - 0s - loss: 30.4633
Epoch 56/100
721/721 - 0s - loss: 30.4355
Epoch 57/100
721/721 - 0s - loss: 30.3703
Epoch 58/100
721/721 - 0s - loss: 30.3995
Epoch 59/100
721/721 - 0s - loss: 30.4133
Epoch 60/100
721/721 - 0s - loss: 30.3762
Epoch 61/100
721/721 - 0s - loss: 30.3681
Epoch 62/100
721/721 - 0s - loss: 30.3776
Epoch 63/100
721/721 - 0s - loss: 30.3397
Epoch 64/100
721/721 - 0s - loss: 30.3738
Epoch 65/100
721/721 - 0s - loss: 30.3188
Epoch 66/100
721/721 - 0s - loss: 30.3806
Epoch 67/100
721/721 - 0s - loss: 30.3666
Epoch 68/100
721/721 - 0s - loss: 30.3890
Epoch 69/100
721/721 - 0s - loss: 30.2641
Epoch 70/100
721/721 - 0s - loss: 30.3294
Epoch 71/100
721/721 - 0s - loss: 30.3184
Epoch 72/100
721/721 - 0s - loss: 30.2775
Epoch 73/100
721/721 - 0s - loss: 30.2811
Epoch 74/100
721/721 - 0s - loss: 

Epoch 46/100
721/721 - 0s - loss: 30.2039
Epoch 47/100
721/721 - 0s - loss: 30.1314
Epoch 48/100
721/721 - 0s - loss: 30.1501
Epoch 49/100
721/721 - 0s - loss: 30.1671
Epoch 50/100
721/721 - 0s - loss: 30.1279
Epoch 51/100
721/721 - 0s - loss: 30.1198
Epoch 52/100
721/721 - 0s - loss: 30.1592
Epoch 53/100
721/721 - 0s - loss: 30.0940
Epoch 54/100
721/721 - 0s - loss: 30.1143
Epoch 55/100
721/721 - 0s - loss: 30.1722
Epoch 56/100
721/721 - 0s - loss: 30.0851
Epoch 57/100
721/721 - 0s - loss: 30.0798
Epoch 58/100
721/721 - 0s - loss: 30.0973
Epoch 59/100
721/721 - 0s - loss: 30.1159
Epoch 60/100
721/721 - 0s - loss: 30.0489
Epoch 61/100
721/721 - 0s - loss: 30.0790
Epoch 62/100
721/721 - 0s - loss: 30.0792
Epoch 63/100
721/721 - 0s - loss: 30.0410
Epoch 64/100
721/721 - 0s - loss: 30.0212
Epoch 65/100
721/721 - 0s - loss: 30.0074
Epoch 66/100
721/721 - 0s - loss: 29.9978
Epoch 67/100
721/721 - 0s - loss: 30.0436
Epoch 68/100
721/721 - 0s - loss: 30.0368
Epoch 69/100
721/721 - 0s - loss: 

Epoch 41/100
721/721 - 0s - loss: 27.8006
Epoch 42/100
721/721 - 0s - loss: 27.7296
Epoch 43/100
721/721 - 0s - loss: 27.7765
Epoch 44/100
721/721 - 0s - loss: 27.7100
Epoch 45/100
721/721 - 0s - loss: 27.7122
Epoch 46/100
721/721 - 0s - loss: 27.6993
Epoch 47/100
721/721 - 0s - loss: 27.7064
Epoch 48/100
721/721 - 0s - loss: 27.7554
Epoch 49/100
721/721 - 0s - loss: 27.6507
Epoch 50/100
721/721 - 0s - loss: 27.6721
Epoch 51/100
721/721 - 0s - loss: 27.6804
Epoch 52/100
721/721 - 0s - loss: 27.6464
Epoch 53/100
721/721 - 0s - loss: 27.6098
Epoch 54/100
721/721 - 0s - loss: 27.6545
Epoch 55/100
721/721 - 0s - loss: 27.6467
Epoch 56/100
721/721 - 0s - loss: 27.5826
Epoch 57/100
721/721 - 0s - loss: 27.6115
Epoch 58/100
721/721 - 0s - loss: 27.5809
Epoch 59/100
721/721 - 0s - loss: 27.6378
Epoch 60/100
721/721 - 0s - loss: 27.5862
Epoch 61/100
721/721 - 0s - loss: 27.6050
Epoch 62/100
721/721 - 0s - loss: 27.5777
Epoch 63/100
721/721 - 0s - loss: 27.5817
Epoch 64/100
721/721 - 0s - loss: 

Epoch 36/100
721/721 - 0s - loss: 26.6107
Epoch 37/100
721/721 - 0s - loss: 26.6232
Epoch 38/100
721/721 - 0s - loss: 26.5800
Epoch 39/100
721/721 - 0s - loss: 26.5693
Epoch 40/100
721/721 - 0s - loss: 26.5960
Epoch 41/100
721/721 - 0s - loss: 26.5436
Epoch 42/100
721/721 - 0s - loss: 26.5874
Epoch 43/100
721/721 - 0s - loss: 26.5303
Epoch 44/100
721/721 - 0s - loss: 26.5553
Epoch 45/100
721/721 - 0s - loss: 26.5154
Epoch 46/100
721/721 - 0s - loss: 26.5602
Epoch 47/100
721/721 - 0s - loss: 26.4981
Epoch 48/100
721/721 - 0s - loss: 26.5447
Epoch 49/100
721/721 - 0s - loss: 26.4802
Epoch 50/100
721/721 - 0s - loss: 26.4977
Epoch 51/100
721/721 - 0s - loss: 26.5335
Epoch 52/100
721/721 - 0s - loss: 26.4637
Epoch 53/100
721/721 - 0s - loss: 26.4475
Epoch 54/100
721/721 - 0s - loss: 26.4270
Epoch 55/100
721/721 - 0s - loss: 26.4403
Epoch 56/100
721/721 - 0s - loss: 26.4173
Epoch 57/100
721/721 - 0s - loss: 26.4775
Epoch 58/100
721/721 - 0s - loss: 26.4352
Epoch 59/100
721/721 - 0s - loss: 

Epoch 31/100
721/721 - 0s - loss: 27.5066
Epoch 32/100
721/721 - 0s - loss: 27.5101
Epoch 33/100
721/721 - 0s - loss: 27.4722
Epoch 34/100
721/721 - 0s - loss: 27.4165
Epoch 35/100
721/721 - 0s - loss: 27.4311
Epoch 36/100
721/721 - 0s - loss: 27.4680
Epoch 37/100
721/721 - 0s - loss: 27.3998
Epoch 38/100
721/721 - 0s - loss: 27.5112
Epoch 39/100
721/721 - 0s - loss: 27.4468
Epoch 40/100
721/721 - 0s - loss: 27.3906
Epoch 41/100
721/721 - 0s - loss: 27.3699
Epoch 42/100
721/721 - 0s - loss: 27.3846
Epoch 43/100
721/721 - 0s - loss: 27.3540
Epoch 44/100
721/721 - 0s - loss: 27.3643
Epoch 45/100
721/721 - 0s - loss: 27.3503
Epoch 46/100
721/721 - 0s - loss: 27.3505
Epoch 47/100
721/721 - 0s - loss: 27.3294
Epoch 48/100
721/721 - 0s - loss: 27.3791
Epoch 49/100
721/721 - 0s - loss: 27.3118
Epoch 50/100
721/721 - 0s - loss: 27.3071
Epoch 51/100
721/721 - 0s - loss: 27.3425
Epoch 52/100
721/721 - 0s - loss: 27.3117
Epoch 53/100
721/721 - 0s - loss: 27.3646
Epoch 54/100
721/721 - 0s - loss: 

Epoch 26/100
721/721 - 0s - loss: 28.5852
Epoch 27/100
721/721 - 0s - loss: 28.5871
Epoch 28/100
721/721 - 0s - loss: 28.5963
Epoch 29/100
721/721 - 0s - loss: 28.5429
Epoch 30/100
721/721 - 0s - loss: 28.5850
Epoch 31/100
721/721 - 0s - loss: 28.5283
Epoch 32/100
721/721 - 0s - loss: 28.6101
Epoch 33/100
721/721 - 0s - loss: 28.5763
Epoch 34/100
721/721 - 0s - loss: 28.5143
Epoch 35/100
721/721 - 0s - loss: 28.5330
Epoch 36/100
721/721 - 0s - loss: 28.5307
Epoch 37/100
721/721 - 0s - loss: 28.4790
Epoch 38/100
721/721 - 0s - loss: 28.5135
Epoch 39/100
721/721 - 0s - loss: 28.5383
Epoch 40/100
721/721 - 0s - loss: 28.6319
Epoch 41/100
721/721 - 0s - loss: 28.5492
Epoch 42/100
721/721 - 0s - loss: 28.4704
Epoch 43/100
721/721 - 0s - loss: 28.5121
Epoch 44/100
721/721 - 0s - loss: 28.4129
Epoch 45/100
721/721 - 0s - loss: 28.4618
Epoch 46/100
721/721 - 0s - loss: 28.4842
Epoch 47/100
721/721 - 0s - loss: 28.5855
Epoch 48/100
721/721 - 0s - loss: 28.4666
Epoch 49/100
721/721 - 0s - loss: 

Epoch 21/100
721/721 - 0s - loss: 26.7626
Epoch 22/100
721/721 - 0s - loss: 26.6914
Epoch 23/100
721/721 - 0s - loss: 26.6655
Epoch 24/100
721/721 - 0s - loss: 26.7773
Epoch 25/100
721/721 - 0s - loss: 26.6527
Epoch 26/100
721/721 - 0s - loss: 26.6386
Epoch 27/100
721/721 - 0s - loss: 26.6249
Epoch 28/100
721/721 - 0s - loss: 26.5828
Epoch 29/100
721/721 - 0s - loss: 26.5971
Epoch 30/100
721/721 - 0s - loss: 26.5857
Epoch 31/100
721/721 - 0s - loss: 26.6098
Epoch 32/100
721/721 - 0s - loss: 26.5927
Epoch 33/100
721/721 - 0s - loss: 26.6236
Epoch 34/100
721/721 - 0s - loss: 26.6166
Epoch 35/100
721/721 - 0s - loss: 26.5811
Epoch 36/100
721/721 - 0s - loss: 26.6453
Epoch 37/100
721/721 - 0s - loss: 26.6027
Epoch 38/100
721/721 - 0s - loss: 26.5351
Epoch 39/100
721/721 - 0s - loss: 26.5550
Epoch 40/100
721/721 - 0s - loss: 26.5163
Epoch 41/100
721/721 - 0s - loss: 26.5261
Epoch 42/100
721/721 - 0s - loss: 26.5284
Epoch 43/100
721/721 - 0s - loss: 26.5025
Epoch 44/100
721/721 - 0s - loss: 

Epoch 16/100
721/721 - 0s - loss: 28.2427
Epoch 17/100
721/721 - 0s - loss: 28.2013
Epoch 18/100
721/721 - 0s - loss: 28.2752
Epoch 19/100
721/721 - 0s - loss: 28.2004
Epoch 20/100
721/721 - 0s - loss: 28.1954
Epoch 21/100
721/721 - 0s - loss: 28.1458
Epoch 22/100
721/721 - 0s - loss: 28.1954
Epoch 23/100
721/721 - 0s - loss: 28.1884
Epoch 24/100
721/721 - 0s - loss: 28.1480
Epoch 25/100
721/721 - 0s - loss: 28.1718
Epoch 26/100
721/721 - 0s - loss: 28.1579
Epoch 27/100
721/721 - 0s - loss: 28.2677
Epoch 28/100
721/721 - 0s - loss: 28.1487
Epoch 29/100
721/721 - 0s - loss: 28.1620
Epoch 30/100
721/721 - 0s - loss: 28.1077
Epoch 31/100
721/721 - 0s - loss: 28.1008
Epoch 32/100
721/721 - 0s - loss: 28.0920
Epoch 33/100
721/721 - 0s - loss: 28.0974
Epoch 34/100
721/721 - 0s - loss: 28.0665
Epoch 35/100
721/721 - 0s - loss: 28.0836
Epoch 36/100
721/721 - 0s - loss: 28.0563
Epoch 37/100
721/721 - 0s - loss: 28.0660
Epoch 38/100
721/721 - 0s - loss: 28.0714
Epoch 39/100
721/721 - 0s - loss: 

Epoch 11/100
721/721 - 0s - loss: 28.7387
Epoch 12/100
721/721 - 0s - loss: 28.7551
Epoch 13/100
721/721 - 0s - loss: 28.6782
Epoch 14/100
721/721 - 0s - loss: 28.6554
Epoch 15/100
721/721 - 0s - loss: 28.6412
Epoch 16/100
721/721 - 0s - loss: 28.6247
Epoch 17/100
721/721 - 0s - loss: 28.6000
Epoch 18/100
721/721 - 0s - loss: 28.6277
Epoch 19/100
721/721 - 0s - loss: 28.5830
Epoch 20/100
721/721 - 0s - loss: 28.5932
Epoch 21/100
721/721 - 0s - loss: 28.5872
Epoch 22/100
721/721 - 0s - loss: 28.5631
Epoch 23/100
721/721 - 0s - loss: 28.5253
Epoch 24/100
721/721 - 0s - loss: 28.5132
Epoch 25/100
721/721 - 0s - loss: 28.4912
Epoch 26/100
721/721 - 0s - loss: 28.4728
Epoch 27/100
721/721 - 0s - loss: 28.5143
Epoch 28/100
721/721 - 0s - loss: 28.4979
Epoch 29/100
721/721 - 0s - loss: 28.4905
Epoch 30/100
721/721 - 0s - loss: 28.4551
Epoch 31/100
721/721 - 0s - loss: 28.4490
Epoch 32/100
721/721 - 0s - loss: 28.4839
Epoch 33/100
721/721 - 0s - loss: 28.5402
Epoch 34/100
721/721 - 0s - loss: 

Epoch 6/100
721/721 - 0s - loss: 26.5899
Epoch 7/100
721/721 - 0s - loss: 26.5663
Epoch 8/100
721/721 - 0s - loss: 26.5249
Epoch 9/100
721/721 - 0s - loss: 26.4435
Epoch 10/100
721/721 - 0s - loss: 26.3893
Epoch 11/100
721/721 - 0s - loss: 26.4118
Epoch 12/100
721/721 - 0s - loss: 26.3568
Epoch 13/100
721/721 - 0s - loss: 26.3256
Epoch 14/100
721/721 - 0s - loss: 26.3261
Epoch 15/100
721/721 - 0s - loss: 26.2834
Epoch 16/100
721/721 - 0s - loss: 26.3831
Epoch 17/100
721/721 - 0s - loss: 26.2836
Epoch 18/100
721/721 - 0s - loss: 26.2415
Epoch 19/100
721/721 - 0s - loss: 26.2434
Epoch 20/100
721/721 - 0s - loss: 26.2501
Epoch 21/100
721/721 - 0s - loss: 26.1996
Epoch 22/100
721/721 - 0s - loss: 26.1965
Epoch 23/100
721/721 - 0s - loss: 26.3159
Epoch 24/100
721/721 - 0s - loss: 26.1733
Epoch 25/100
721/721 - 0s - loss: 26.1746
Epoch 26/100
721/721 - 0s - loss: 26.1544
Epoch 27/100
721/721 - 0s - loss: 26.1482
Epoch 28/100
721/721 - 0s - loss: 26.1388
Epoch 29/100
721/721 - 0s - loss: 26.1

Train on 721 samples
Epoch 1/100
721/721 - 0s - loss: 25.7707
Epoch 2/100
721/721 - 0s - loss: 25.6063
Epoch 3/100
721/721 - 0s - loss: 25.4966
Epoch 4/100
721/721 - 0s - loss: 25.4203
Epoch 5/100
721/721 - 0s - loss: 25.3624
Epoch 6/100
721/721 - 0s - loss: 25.3065
Epoch 7/100
721/721 - 0s - loss: 25.2968
Epoch 8/100
721/721 - 0s - loss: 25.2522
Epoch 9/100
721/721 - 0s - loss: 25.2872
Epoch 10/100
721/721 - 0s - loss: 25.2425
Epoch 11/100
721/721 - 0s - loss: 25.1852
Epoch 12/100
721/721 - 0s - loss: 25.2355
Epoch 13/100
721/721 - 0s - loss: 25.1459
Epoch 14/100
721/721 - 0s - loss: 25.1699
Epoch 15/100
721/721 - 0s - loss: 25.1323
Epoch 16/100
721/721 - 0s - loss: 25.1227
Epoch 17/100
721/721 - 0s - loss: 25.1132
Epoch 18/100
721/721 - 0s - loss: 25.0597
Epoch 19/100
721/721 - 0s - loss: 25.0782
Epoch 20/100
721/721 - 0s - loss: 25.0932
Epoch 21/100
721/721 - 0s - loss: 25.0544
Epoch 22/100
721/721 - 0s - loss: 25.0439
Epoch 23/100
721/721 - 0s - loss: 25.0400
Epoch 24/100
721/721 -

Epoch 96/100
721/721 - 0s - loss: 26.0207
Epoch 97/100
721/721 - 0s - loss: 25.9908
Epoch 98/100
721/721 - 0s - loss: 26.0033
Epoch 99/100
721/721 - 0s - loss: 26.0073
Epoch 100/100
721/721 - 0s - loss: 25.9914
Train on 721 samples
Epoch 1/100
721/721 - 0s - loss: 27.1676
Epoch 2/100
721/721 - 0s - loss: 26.9934
Epoch 3/100
721/721 - 0s - loss: 26.9453
Epoch 4/100
721/721 - 0s - loss: 26.9149
Epoch 5/100
721/721 - 0s - loss: 26.9524
Epoch 6/100
721/721 - 0s - loss: 26.8489
Epoch 7/100
721/721 - 0s - loss: 26.8418
Epoch 8/100
721/721 - 0s - loss: 26.8693
Epoch 9/100
721/721 - 0s - loss: 26.8368
Epoch 10/100
721/721 - 0s - loss: 26.8562
Epoch 11/100
721/721 - 0s - loss: 26.9076
Epoch 12/100
721/721 - 0s - loss: 26.7936
Epoch 13/100
721/721 - 0s - loss: 26.8075
Epoch 14/100
721/721 - 0s - loss: 26.7592
Epoch 15/100
721/721 - 0s - loss: 26.7762
Epoch 16/100
721/721 - 0s - loss: 26.7228
Epoch 17/100
721/721 - 0s - loss: 26.7398
Epoch 18/100
721/721 - 0s - loss: 26.6752
Epoch 19/100
721/721 

Epoch 91/100
721/721 - 0s - loss: 27.1482
Epoch 92/100
721/721 - 0s - loss: 27.1687
Epoch 93/100
721/721 - 0s - loss: 27.1338
Epoch 94/100
721/721 - 0s - loss: 27.1268
Epoch 95/100
721/721 - 0s - loss: 27.1285
Epoch 96/100
721/721 - 0s - loss: 27.1299
Epoch 97/100
721/721 - 0s - loss: 27.1536
Epoch 98/100
721/721 - 0s - loss: 27.1436
Epoch 99/100
721/721 - 0s - loss: 27.1349
Epoch 100/100
721/721 - 0s - loss: 27.1548
Train on 721 samples
Epoch 1/100
721/721 - 0s - loss: 28.4296
Epoch 2/100
721/721 - 0s - loss: 28.2828
Epoch 3/100
721/721 - 0s - loss: 28.2332
Epoch 4/100
721/721 - 0s - loss: 28.2071
Epoch 5/100
721/721 - 0s - loss: 28.1923
Epoch 6/100
721/721 - 0s - loss: 28.1716
Epoch 7/100
721/721 - 0s - loss: 28.1115
Epoch 8/100
721/721 - 0s - loss: 28.0995
Epoch 9/100
721/721 - 0s - loss: 28.0824
Epoch 10/100
721/721 - 0s - loss: 28.1342
Epoch 11/100
721/721 - 0s - loss: 28.0401
Epoch 12/100
721/721 - 0s - loss: 28.0472
Epoch 13/100
721/721 - 0s - loss: 28.1347
Epoch 14/100
721/721 

Epoch 86/100
721/721 - 0s - loss: 25.9566
Epoch 87/100
721/721 - 0s - loss: 25.8480
Epoch 88/100
721/721 - 0s - loss: 25.9132
Epoch 89/100
721/721 - 0s - loss: 25.8857
Epoch 90/100
721/721 - 0s - loss: 25.8742
Epoch 91/100
721/721 - 0s - loss: 26.0101
Epoch 92/100
721/721 - 0s - loss: 25.8978
Epoch 93/100
721/721 - 0s - loss: 26.0452
Epoch 94/100
721/721 - 0s - loss: 25.9275
Epoch 95/100
721/721 - 0s - loss: 25.8777
Epoch 96/100
721/721 - 0s - loss: 25.8794
Epoch 97/100
721/721 - 0s - loss: 25.8947
Epoch 98/100
721/721 - 0s - loss: 25.9674
Epoch 99/100
721/721 - 0s - loss: 25.8572
Epoch 100/100
721/721 - 0s - loss: 25.8658
Train on 721 samples
Epoch 1/100
721/721 - 0s - loss: 25.6048
Epoch 2/100
721/721 - 0s - loss: 25.6788
Epoch 3/100
721/721 - 0s - loss: 25.4437
Epoch 4/100
721/721 - 0s - loss: 25.4288
Epoch 5/100
721/721 - 0s - loss: 25.3967
Epoch 6/100
721/721 - 0s - loss: 25.4014
Epoch 7/100
721/721 - 0s - loss: 25.2649
Epoch 8/100
721/721 - 0s - loss: 25.3127
Epoch 9/100
721/721 

Epoch 81/100
721/721 - 0s - loss: 25.5829
Epoch 82/100
721/721 - 0s - loss: 25.5517
Epoch 83/100
721/721 - 0s - loss: 25.5529
Epoch 84/100
721/721 - 0s - loss: 25.5388
Epoch 85/100
721/721 - 0s - loss: 25.5803
Epoch 86/100
721/721 - 0s - loss: 25.5411
Epoch 87/100
721/721 - 0s - loss: 25.5061
Epoch 88/100
721/721 - 0s - loss: 25.5132
Epoch 89/100
721/721 - 0s - loss: 25.5977
Epoch 90/100
721/721 - 0s - loss: 25.5217
Epoch 91/100
721/721 - 0s - loss: 25.5404
Epoch 92/100
721/721 - 0s - loss: 25.6206
Epoch 93/100
721/721 - 0s - loss: 25.4994
Epoch 94/100
721/721 - 0s - loss: 25.5284
Epoch 95/100
721/721 - 0s - loss: 25.5395
Epoch 96/100
721/721 - 0s - loss: 25.5290
Epoch 97/100
721/721 - 0s - loss: 25.5537
Epoch 98/100
721/721 - 0s - loss: 25.5458
Epoch 99/100
721/721 - 0s - loss: 25.5048
Epoch 100/100
721/721 - 0s - loss: 25.5089
Train on 721 samples
Epoch 1/100
721/721 - 0s - loss: 25.8482
Epoch 2/100
721/721 - 0s - loss: 25.7442
Epoch 3/100
721/721 - 0s - loss: 25.6971
Epoch 4/100
721

Epoch 76/100
721/721 - 0s - loss: 26.7278
Epoch 77/100
721/721 - 0s - loss: 26.8190
Epoch 78/100
721/721 - 0s - loss: 26.7424
Epoch 79/100
721/721 - 0s - loss: 26.8418
Epoch 80/100
721/721 - 0s - loss: 26.7325
Epoch 81/100
721/721 - 0s - loss: 26.7285
Epoch 82/100
721/721 - 0s - loss: 26.7675
Epoch 83/100
721/721 - 0s - loss: 26.7675
Epoch 84/100
721/721 - 0s - loss: 26.6799
Epoch 85/100
721/721 - 0s - loss: 26.7679
Epoch 86/100
721/721 - 0s - loss: 26.6984
Epoch 87/100
721/721 - 0s - loss: 26.7329
Epoch 88/100
721/721 - 0s - loss: 26.7673
Epoch 89/100
721/721 - 0s - loss: 26.7238
Epoch 90/100
721/721 - 0s - loss: 26.6960
Epoch 91/100
721/721 - 0s - loss: 26.7149
Epoch 92/100
721/721 - 0s - loss: 26.7020
Epoch 93/100
721/721 - 0s - loss: 26.7856
Epoch 94/100
721/721 - 0s - loss: 26.7400
Epoch 95/100
721/721 - 0s - loss: 26.7175
Epoch 96/100
721/721 - 0s - loss: 26.6973
Epoch 97/100
721/721 - 0s - loss: 26.7392
Epoch 98/100
721/721 - 0s - loss: 26.6720
Epoch 99/100
721/721 - 0s - loss: 

Epoch 71/100
721/721 - 0s - loss: 24.2766
Epoch 72/100
721/721 - 0s - loss: 24.2729
Epoch 73/100
721/721 - 0s - loss: 24.2684
Epoch 74/100
721/721 - 0s - loss: 24.3195
Epoch 75/100
721/721 - 0s - loss: 24.1865
Epoch 76/100
721/721 - 0s - loss: 24.3189
Epoch 77/100
721/721 - 0s - loss: 24.1923
Epoch 78/100
721/721 - 0s - loss: 24.2103
Epoch 79/100
721/721 - 0s - loss: 24.2444
Epoch 80/100
721/721 - 0s - loss: 24.1711
Epoch 81/100
721/721 - 0s - loss: 24.2005
Epoch 82/100
721/721 - 0s - loss: 24.2045
Epoch 83/100
721/721 - 0s - loss: 24.1765
Epoch 84/100
721/721 - 0s - loss: 24.2113
Epoch 85/100
721/721 - 0s - loss: 24.1588
Epoch 86/100
721/721 - 0s - loss: 24.1373
Epoch 87/100
721/721 - 0s - loss: 24.2139
Epoch 88/100
721/721 - 0s - loss: 24.2019
Epoch 89/100
721/721 - 0s - loss: 24.1487
Epoch 90/100
721/721 - 0s - loss: 24.1819
Epoch 91/100
721/721 - 0s - loss: 24.1223
Epoch 92/100
721/721 - 0s - loss: 24.1680
Epoch 93/100
721/721 - 0s - loss: 24.1489
Epoch 94/100
721/721 - 0s - loss: 

In [105]:
print(mse_list)

[187.91231203891414, 96.17614909894313, 55.118918717351555, 50.49760109480258, 41.107478935459596, 39.50779624070217, 38.15503855861852, 35.86428123059918, 38.791068412915045, 40.14471660998352, 33.166374716729834, 30.044926249136463, 37.15539826258989, 31.311339716026122, 32.09990172803143, 37.89460615929648, 36.72393449738827, 30.328392797680234, 33.46772495082088, 28.783058065340253, 28.802151954868524, 33.66682762858739, 31.108922749255616, 35.10860082337091, 27.592397471130706, 31.521422075461626, 32.03607203880258, 28.14002689044952, 29.98811110599913, 32.158376649191, 27.504421507124107, 27.558214007556575, 29.580014929684104, 26.691959156343632, 34.19465672798526, 31.806955211312257, 27.792252308345297, 33.08402150064319, 29.755165261787205, 29.13117365155605, 26.45995777674994, 26.23791980445853, 29.51280592926003, 32.15734336461292, 30.289781064632837, 31.30057978929733, 27.401166715367726, 26.651258699446146, 32.73452754841693, 31.61359501105688]


In [106]:
# Calculate mean and standard deviation for the stored mse values

mse_mean = np.array(mse_list).mean()
mse_std = np.array(mse_list).std()

print(f'The mean of the mean squared errors is {mse_mean}')
print(f'The standard deviation of the mean squared errors is {mse_std}')

The mean of the mean squared errors is 37.11663394868167
The standard deviation of the mean squared errors is 23.958755060749166


#### The mean of the mean squared errors compared to Part B has decreased from 50.31 to 37.12

##  Part D

### Increase the number of hidden layers 

In [107]:
data = pd.read_csv('https://cocl.us/concrete_data')
data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [108]:
# Splitting the data into predictors and target

predictors = data[data.columns[data.columns != 'Strength']]
target = data['Strength']
print(target.shape)

(1030,)


In [109]:
# Normalize the data

predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [110]:
#Let's save the number of predictors to n_cols since we will need this number when building our network.

n_cols = predictors_norm.shape[1]
n_cols

8

In [111]:
# Three hidden layers, each of 10 nodes and ReLU activation function.

model = tf.keras.Sequential([
                        tf.keras.layers.Dense(10,activation="relu",input_shape=(n_cols,)),
                        tf.keras.layers.Dense(10,activation="relu"),
                        tf.keras.layers.Dense(10,activation="relu"),
                        tf.keras.layers.Dense(1)
                        ])

model.compile(optimizer='adam', loss='mean_squared_error')

In [112]:
mse_list = []
for i in range(1,51):
    
    # Randomly split the data into a training and test sets by holding 30% of the data for testing.
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm,target,test_size=0.3)
    
    # Fit the model and train for 50 epochs
    model.fit(X_train, y_train, epochs=50, verbose=2)
    
    # Predict the target
    y_pred = model.predict(X_test)
    
    # Calculate mean squared error and store it in a list
    from sklearn.metrics import mean_squared_error
    mse = mean_squared_error(y_test,y_pred)
    mse_list.append(mse)

Train on 721 samples
Epoch 1/50
721/721 - 1s - loss: 1608.2783
Epoch 2/50
721/721 - 0s - loss: 1586.5925
Epoch 3/50
721/721 - 0s - loss: 1564.0936
Epoch 4/50
721/721 - 0s - loss: 1533.0532
Epoch 5/50
721/721 - 0s - loss: 1485.7330
Epoch 6/50
721/721 - 0s - loss: 1416.8052
Epoch 7/50
721/721 - 0s - loss: 1319.3675
Epoch 8/50
721/721 - 0s - loss: 1192.3471
Epoch 9/50
721/721 - 0s - loss: 1042.7737
Epoch 10/50
721/721 - 0s - loss: 883.4479
Epoch 11/50
721/721 - 0s - loss: 739.3194
Epoch 12/50
721/721 - 0s - loss: 614.0512
Epoch 13/50
721/721 - 0s - loss: 507.9493
Epoch 14/50
721/721 - 0s - loss: 416.4486
Epoch 15/50
721/721 - 0s - loss: 344.5363
Epoch 16/50
721/721 - 0s - loss: 290.4005
Epoch 17/50
721/721 - 0s - loss: 255.0837
Epoch 18/50
721/721 - 0s - loss: 231.7412
Epoch 19/50
721/721 - 0s - loss: 215.1301
Epoch 20/50
721/721 - 0s - loss: 204.8024
Epoch 21/50
721/721 - 0s - loss: 196.5021
Epoch 22/50
721/721 - 0s - loss: 190.4858
Epoch 23/50
721/721 - 0s - loss: 185.1956
Epoch 24/50
7

Epoch 46/50
721/721 - 0s - loss: 79.4545
Epoch 47/50
721/721 - 0s - loss: 79.4134
Epoch 48/50
721/721 - 0s - loss: 78.7264
Epoch 49/50
721/721 - 0s - loss: 78.4761
Epoch 50/50
721/721 - 0s - loss: 78.8338
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 81.2893
Epoch 2/50
721/721 - 0s - loss: 80.4220
Epoch 3/50
721/721 - 0s - loss: 79.9850
Epoch 4/50
721/721 - 0s - loss: 79.6200
Epoch 5/50
721/721 - 0s - loss: 79.1378
Epoch 6/50
721/721 - 0s - loss: 78.8207
Epoch 7/50
721/721 - 0s - loss: 78.4623
Epoch 8/50
721/721 - 0s - loss: 78.2770
Epoch 9/50
721/721 - 0s - loss: 77.9233
Epoch 10/50
721/721 - 0s - loss: 77.6355
Epoch 11/50
721/721 - 0s - loss: 77.3181
Epoch 12/50
721/721 - 0s - loss: 77.0561
Epoch 13/50
721/721 - 0s - loss: 76.6131
Epoch 14/50
721/721 - 0s - loss: 76.3785
Epoch 15/50
721/721 - 0s - loss: 76.1420
Epoch 16/50
721/721 - 0s - loss: 75.6969
Epoch 17/50
721/721 - 0s - loss: 75.5755
Epoch 18/50
721/721 - 0s - loss: 75.3190
Epoch 19/50
721/721 - 0s - loss: 75.0964
Epoc

Epoch 45/50
721/721 - 0s - loss: 37.8565
Epoch 46/50
721/721 - 0s - loss: 37.7529
Epoch 47/50
721/721 - 0s - loss: 37.7590
Epoch 48/50
721/721 - 0s - loss: 37.2966
Epoch 49/50
721/721 - 0s - loss: 37.6822
Epoch 50/50
721/721 - 0s - loss: 37.2462
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 42.8374
Epoch 2/50
721/721 - 0s - loss: 41.5066
Epoch 3/50
721/721 - 0s - loss: 41.6511
Epoch 4/50
721/721 - 0s - loss: 41.5275
Epoch 5/50
721/721 - 0s - loss: 41.8427
Epoch 6/50
721/721 - 0s - loss: 40.8301
Epoch 7/50
721/721 - 0s - loss: 40.8231
Epoch 8/50
721/721 - 0s - loss: 40.7142
Epoch 9/50
721/721 - 0s - loss: 40.6499
Epoch 10/50
721/721 - 0s - loss: 40.4745
Epoch 11/50
721/721 - 0s - loss: 40.2684
Epoch 12/50
721/721 - 0s - loss: 40.4664
Epoch 13/50
721/721 - 0s - loss: 40.4893
Epoch 14/50
721/721 - 0s - loss: 39.7159
Epoch 15/50
721/721 - 0s - loss: 39.7600
Epoch 16/50
721/721 - 0s - loss: 39.6227
Epoch 17/50
721/721 - 0s - loss: 40.0639
Epoch 18/50
721/721 - 0s - loss: 39.5090
Epoc

Epoch 44/50
721/721 - 0s - loss: 29.0964
Epoch 45/50
721/721 - 0s - loss: 29.1973
Epoch 46/50
721/721 - 0s - loss: 28.6571
Epoch 47/50
721/721 - 0s - loss: 28.7839
Epoch 48/50
721/721 - 0s - loss: 28.8805
Epoch 49/50
721/721 - 0s - loss: 28.8464
Epoch 50/50
721/721 - 0s - loss: 28.6636
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 29.2802
Epoch 2/50
721/721 - 0s - loss: 28.9822
Epoch 3/50
721/721 - 0s - loss: 28.4986
Epoch 4/50
721/721 - 0s - loss: 28.3466
Epoch 5/50
721/721 - 0s - loss: 28.1425
Epoch 6/50
721/721 - 0s - loss: 28.6777
Epoch 7/50
721/721 - 0s - loss: 28.5817
Epoch 8/50
721/721 - 0s - loss: 28.2273
Epoch 9/50
721/721 - 0s - loss: 28.2162
Epoch 10/50
721/721 - 0s - loss: 27.9196
Epoch 11/50
721/721 - 0s - loss: 28.2003
Epoch 12/50
721/721 - 0s - loss: 28.4534
Epoch 13/50
721/721 - 0s - loss: 27.7766
Epoch 14/50
721/721 - 0s - loss: 28.4294
Epoch 15/50
721/721 - 0s - loss: 27.9371
Epoch 16/50
721/721 - 0s - loss: 27.6744
Epoch 17/50
721/721 - 0s - loss: 27.8968
Epoc

Epoch 43/50
721/721 - 0s - loss: 26.6656
Epoch 44/50
721/721 - 0s - loss: 26.6990
Epoch 45/50
721/721 - 0s - loss: 26.6756
Epoch 46/50
721/721 - 0s - loss: 26.7660
Epoch 47/50
721/721 - 0s - loss: 26.5640
Epoch 48/50
721/721 - 0s - loss: 26.3500
Epoch 49/50
721/721 - 0s - loss: 26.8338
Epoch 50/50
721/721 - 0s - loss: 27.1078
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 28.1026
Epoch 2/50
721/721 - 0s - loss: 26.9324
Epoch 3/50
721/721 - 0s - loss: 26.6751
Epoch 4/50
721/721 - 0s - loss: 26.8232
Epoch 5/50
721/721 - 0s - loss: 26.6836
Epoch 6/50
721/721 - 0s - loss: 26.4603
Epoch 7/50
721/721 - 0s - loss: 27.1196
Epoch 8/50
721/721 - 0s - loss: 26.9210
Epoch 9/50
721/721 - 0s - loss: 26.7747
Epoch 10/50
721/721 - 0s - loss: 26.6397
Epoch 11/50
721/721 - 0s - loss: 26.6618
Epoch 12/50
721/721 - 0s - loss: 26.3726
Epoch 13/50
721/721 - 0s - loss: 26.3364
Epoch 14/50
721/721 - 0s - loss: 26.4170
Epoch 15/50
721/721 - 0s - loss: 26.2534
Epoch 16/50
721/721 - 0s - loss: 26.2768
Epoc

Epoch 42/50
721/721 - 0s - loss: 24.0820
Epoch 43/50
721/721 - 0s - loss: 24.1282
Epoch 44/50
721/721 - 0s - loss: 24.2095
Epoch 45/50
721/721 - 0s - loss: 24.1563
Epoch 46/50
721/721 - 0s - loss: 24.1458
Epoch 47/50
721/721 - 0s - loss: 24.2650
Epoch 48/50
721/721 - 0s - loss: 24.0795
Epoch 49/50
721/721 - 0s - loss: 24.2746
Epoch 50/50
721/721 - 0s - loss: 24.2662
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 26.8050
Epoch 2/50
721/721 - 0s - loss: 26.9317
Epoch 3/50
721/721 - 0s - loss: 26.7857
Epoch 4/50
721/721 - 0s - loss: 26.7026
Epoch 5/50
721/721 - 0s - loss: 26.2152
Epoch 6/50
721/721 - 0s - loss: 26.1793
Epoch 7/50
721/721 - 0s - loss: 26.2044
Epoch 8/50
721/721 - 0s - loss: 26.2001
Epoch 9/50
721/721 - 0s - loss: 26.3257
Epoch 10/50
721/721 - 0s - loss: 25.9520
Epoch 11/50
721/721 - 0s - loss: 25.8329
Epoch 12/50
721/721 - 0s - loss: 26.1502
Epoch 13/50
721/721 - 0s - loss: 26.0260
Epoch 14/50
721/721 - 0s - loss: 25.9773
Epoch 15/50
721/721 - 0s - loss: 26.4780
Epoc

Epoch 41/50
721/721 - 0s - loss: 23.4727
Epoch 42/50
721/721 - 0s - loss: 23.2614
Epoch 43/50
721/721 - 0s - loss: 23.2245
Epoch 44/50
721/721 - 0s - loss: 23.4150
Epoch 45/50
721/721 - 0s - loss: 23.2511
Epoch 46/50
721/721 - 0s - loss: 23.0603
Epoch 47/50
721/721 - 0s - loss: 23.2463
Epoch 48/50
721/721 - 0s - loss: 23.3476
Epoch 49/50
721/721 - 0s - loss: 23.1669
Epoch 50/50
721/721 - 0s - loss: 23.4072
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 24.7739
Epoch 2/50
721/721 - 0s - loss: 24.3628
Epoch 3/50
721/721 - 0s - loss: 24.3275
Epoch 4/50
721/721 - 0s - loss: 24.0513
Epoch 5/50
721/721 - 0s - loss: 24.3714
Epoch 6/50
721/721 - 0s - loss: 24.2333
Epoch 7/50
721/721 - 0s - loss: 24.2453
Epoch 8/50
721/721 - 0s - loss: 23.9048
Epoch 9/50
721/721 - 0s - loss: 23.9968
Epoch 10/50
721/721 - 0s - loss: 23.9517
Epoch 11/50
721/721 - 0s - loss: 23.8237
Epoch 12/50
721/721 - 0s - loss: 23.7451
Epoch 13/50
721/721 - 0s - loss: 23.9473
Epoch 14/50
721/721 - 0s - loss: 23.8866
Epoc

Epoch 40/50
721/721 - 0s - loss: 24.1716
Epoch 41/50
721/721 - 0s - loss: 24.1315
Epoch 42/50
721/721 - 0s - loss: 24.4276
Epoch 43/50
721/721 - 0s - loss: 23.8413
Epoch 44/50
721/721 - 0s - loss: 23.9689
Epoch 45/50
721/721 - 0s - loss: 23.9773
Epoch 46/50
721/721 - 0s - loss: 24.1233
Epoch 47/50
721/721 - 0s - loss: 24.5642
Epoch 48/50
721/721 - 0s - loss: 24.5369
Epoch 49/50
721/721 - 0s - loss: 23.8744
Epoch 50/50
721/721 - 0s - loss: 24.2465
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 25.2169
Epoch 2/50
721/721 - 0s - loss: 24.6146
Epoch 3/50
721/721 - 0s - loss: 23.9066
Epoch 4/50
721/721 - 0s - loss: 23.9274
Epoch 5/50
721/721 - 0s - loss: 23.9591
Epoch 6/50
721/721 - 0s - loss: 24.0540
Epoch 7/50
721/721 - 0s - loss: 23.8239
Epoch 8/50
721/721 - 0s - loss: 23.8967
Epoch 9/50
721/721 - 0s - loss: 23.8461
Epoch 10/50
721/721 - 0s - loss: 23.6885
Epoch 11/50
721/721 - 0s - loss: 23.7824
Epoch 12/50
721/721 - 0s - loss: 23.6971
Epoch 13/50
721/721 - 0s - loss: 23.8387
Epoc

Epoch 39/50
721/721 - 0s - loss: 21.9862
Epoch 40/50
721/721 - 0s - loss: 21.7894
Epoch 41/50
721/721 - 0s - loss: 21.6367
Epoch 42/50
721/721 - 0s - loss: 21.8125
Epoch 43/50
721/721 - 0s - loss: 22.0070
Epoch 44/50
721/721 - 0s - loss: 21.7499
Epoch 45/50
721/721 - 0s - loss: 21.8225
Epoch 46/50
721/721 - 0s - loss: 21.9754
Epoch 47/50
721/721 - 0s - loss: 22.0287
Epoch 48/50
721/721 - 0s - loss: 21.6685
Epoch 49/50
721/721 - 0s - loss: 21.5732
Epoch 50/50
721/721 - 0s - loss: 21.5630
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 23.0718
Epoch 2/50
721/721 - 0s - loss: 22.8831
Epoch 3/50
721/721 - 0s - loss: 22.7080
Epoch 4/50
721/721 - 0s - loss: 22.9847
Epoch 5/50
721/721 - 0s - loss: 22.1922
Epoch 6/50
721/721 - 0s - loss: 22.5300
Epoch 7/50
721/721 - 0s - loss: 22.3149
Epoch 8/50
721/721 - 0s - loss: 22.3714
Epoch 9/50
721/721 - 0s - loss: 21.9891
Epoch 10/50
721/721 - 0s - loss: 22.4090
Epoch 11/50
721/721 - 0s - loss: 22.1576
Epoch 12/50
721/721 - 0s - loss: 22.0926
Epoc

Epoch 38/50
721/721 - 0s - loss: 19.8324
Epoch 39/50
721/721 - 0s - loss: 19.4270
Epoch 40/50
721/721 - 0s - loss: 19.7171
Epoch 41/50
721/721 - 0s - loss: 20.0263
Epoch 42/50
721/721 - 0s - loss: 19.7849
Epoch 43/50
721/721 - 0s - loss: 19.3495
Epoch 44/50
721/721 - 0s - loss: 19.1909
Epoch 45/50
721/721 - 0s - loss: 19.5288
Epoch 46/50
721/721 - 0s - loss: 19.1842
Epoch 47/50
721/721 - 0s - loss: 19.3600
Epoch 48/50
721/721 - 0s - loss: 19.2779
Epoch 49/50
721/721 - 0s - loss: 19.7678
Epoch 50/50
721/721 - 0s - loss: 19.3301
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 21.8862
Epoch 2/50
721/721 - 0s - loss: 21.5523
Epoch 3/50
721/721 - 0s - loss: 21.3706
Epoch 4/50
721/721 - 0s - loss: 22.1914
Epoch 5/50
721/721 - 0s - loss: 21.7088
Epoch 6/50
721/721 - 0s - loss: 21.6920
Epoch 7/50
721/721 - 0s - loss: 21.0644
Epoch 8/50
721/721 - 0s - loss: 20.8684
Epoch 9/50
721/721 - 0s - loss: 21.1293
Epoch 10/50
721/721 - 0s - loss: 21.0235
Epoch 11/50
721/721 - 0s - loss: 20.8203
Epoc

Epoch 37/50
721/721 - 0s - loss: 19.2083
Epoch 38/50
721/721 - 0s - loss: 19.1323
Epoch 39/50
721/721 - 0s - loss: 19.1739
Epoch 40/50
721/721 - 0s - loss: 19.1212
Epoch 41/50
721/721 - 0s - loss: 19.2641
Epoch 42/50
721/721 - 0s - loss: 19.2912
Epoch 43/50
721/721 - 0s - loss: 19.5364
Epoch 44/50
721/721 - 0s - loss: 19.2759
Epoch 45/50
721/721 - 0s - loss: 19.1454
Epoch 46/50
721/721 - 0s - loss: 19.3998
Epoch 47/50
721/721 - 0s - loss: 19.3828
Epoch 48/50
721/721 - 0s - loss: 19.0411
Epoch 49/50
721/721 - 0s - loss: 19.0005
Epoch 50/50
721/721 - 0s - loss: 19.1834
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 19.9175
Epoch 2/50
721/721 - 0s - loss: 19.6337
Epoch 3/50
721/721 - 0s - loss: 19.4153
Epoch 4/50
721/721 - 0s - loss: 19.1555
Epoch 5/50
721/721 - 0s - loss: 19.2182
Epoch 6/50
721/721 - 0s - loss: 18.9471
Epoch 7/50
721/721 - 0s - loss: 18.8511
Epoch 8/50
721/721 - 0s - loss: 18.6709
Epoch 9/50
721/721 - 0s - loss: 18.9916
Epoch 10/50
721/721 - 0s - loss: 19.0653
Epoc

Epoch 36/50
721/721 - 0s - loss: 21.3782
Epoch 37/50
721/721 - 0s - loss: 21.0796
Epoch 38/50
721/721 - 0s - loss: 21.0967
Epoch 39/50
721/721 - 0s - loss: 21.9031
Epoch 40/50
721/721 - 0s - loss: 21.7947
Epoch 41/50
721/721 - 0s - loss: 21.0246
Epoch 42/50
721/721 - 0s - loss: 21.1432
Epoch 43/50
721/721 - 0s - loss: 21.1470
Epoch 44/50
721/721 - 0s - loss: 21.3417
Epoch 45/50
721/721 - 0s - loss: 21.4996
Epoch 46/50
721/721 - 0s - loss: 21.0415
Epoch 47/50
721/721 - 0s - loss: 21.5258
Epoch 48/50
721/721 - 0s - loss: 21.1956
Epoch 49/50
721/721 - 0s - loss: 21.9220
Epoch 50/50
721/721 - 0s - loss: 21.2055
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 18.7766
Epoch 2/50
721/721 - 0s - loss: 18.8441
Epoch 3/50
721/721 - 0s - loss: 18.6929
Epoch 4/50
721/721 - 0s - loss: 18.3105
Epoch 5/50
721/721 - 0s - loss: 18.4119
Epoch 6/50
721/721 - 0s - loss: 18.4538
Epoch 7/50
721/721 - 0s - loss: 18.6675
Epoch 8/50
721/721 - 0s - loss: 18.1478
Epoch 9/50
721/721 - 0s - loss: 18.2787
Epoc

Epoch 35/50
721/721 - 0s - loss: 20.0931
Epoch 36/50
721/721 - 0s - loss: 20.6382
Epoch 37/50
721/721 - 0s - loss: 20.1789
Epoch 38/50
721/721 - 0s - loss: 20.1214
Epoch 39/50
721/721 - 0s - loss: 20.0965
Epoch 40/50
721/721 - 0s - loss: 20.7249
Epoch 41/50
721/721 - 0s - loss: 20.3640
Epoch 42/50
721/721 - 0s - loss: 20.2195
Epoch 43/50
721/721 - 0s - loss: 20.2577
Epoch 44/50
721/721 - 0s - loss: 19.7283
Epoch 45/50
721/721 - 0s - loss: 20.0269
Epoch 46/50
721/721 - 0s - loss: 19.7572
Epoch 47/50
721/721 - 0s - loss: 19.9873
Epoch 48/50
721/721 - 0s - loss: 20.1241
Epoch 49/50
721/721 - 0s - loss: 20.8021
Epoch 50/50
721/721 - 0s - loss: 20.0174
Train on 721 samples
Epoch 1/50
721/721 - 0s - loss: 20.5004
Epoch 2/50
721/721 - 0s - loss: 19.7711
Epoch 3/50
721/721 - 0s - loss: 20.0599
Epoch 4/50
721/721 - 0s - loss: 19.6001
Epoch 5/50
721/721 - 0s - loss: 19.8697
Epoch 6/50
721/721 - 0s - loss: 19.6344
Epoch 7/50
721/721 - 0s - loss: 19.5700
Epoch 8/50
721/721 - 0s - loss: 19.4348
Epo

In [113]:
print(mse_list)

[127.38863908890109, 132.61350503343593, 100.18738958480928, 89.3180927927656, 72.0408879951739, 59.573093711923114, 49.929488680009634, 43.11649409120865, 28.7574302624321, 41.5806619367392, 34.33941631216476, 36.68904396999064, 37.54734310769942, 32.93965556229987, 30.259575649177886, 31.998328134669183, 33.256916044494766, 32.26709178716962, 27.994121967213424, 32.3103754600316, 29.33223723496113, 27.874025687359577, 35.97076098436982, 30.609377430816572, 29.304562299088804, 32.387253311639526, 26.8184045898481, 25.327350393374807, 25.724688227376227, 26.975675439045347, 25.502664714983624, 26.003145978270332, 25.6319688192513, 23.47923376741078, 22.354739756219363, 27.196747307043193, 23.887516965059422, 22.918916526677524, 24.799508453740426, 23.402670401851864, 26.935774027717954, 19.390618417130263, 20.044227231960633, 17.069116961384836, 25.21977908395273, 23.393110665984167, 25.26598054248521, 19.683357819758243, 20.658564222148055, 21.0641212980913]


In [114]:
# Calculate mean and standard deviation for the stored mse values

mse_mean = np.array(mse_list).mean()
mse_std = np.array(mse_list).std()

print(f'The mean of the mean squared errors is {mse_mean}')
print(f'The standard deviation of the mean squared errors is {mse_std}')

The mean of the mean squared errors is 36.56667299462622
The standard deviation of the mean squared errors is 24.894828291949928


#### The mean of the mean squared errors compared to Part B has decreased from 50.31 to 36.57